{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dfd2c33a-c73b-4eac-a6ec-d1eec2842c7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "# SRC_DIR environment variable should be the absolute path to the 'multicopy-STR-genotyping' directory\n",
    "sys.path.append(os.environ[\"SRC_DIR\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4214da53-578c-41a3-942c-a21a1838348c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from multicopy_STR_genotyping import file_io\n",
    "\n",
    "sns.set_context(\"poster\")\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "9aea3ada-53fc-45a4-8743-b8b620699841",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>str_id</th>\n",
       "      <th>copy_number</th>\n",
       "      <th>frequencies_gangstr</th>\n",
       "      <th>genotype_gangstr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>simulated_trisomy_21</td>\n",
       "      <td>chr21_5011212</td>\n",
       "      <td>2</td>\n",
       "      <td>{3: 56}</td>\n",
       "      <td>[3, 3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>simulated_trisomy_21</td>\n",
       "      <td>chr21_5013077</td>\n",
       "      <td>2</td>\n",
       "      <td>{6: 50}</td>\n",
       "      <td>[6, 6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>simulated_trisomy_21</td>\n",
       "      <td>chr21_5014892</td>\n",
       "      <td>2</td>\n",
       "      <td>{6: 44}</td>\n",
       "      <td>[6, 6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>simulated_trisomy_21</td>\n",
       "      <td>chr21_5016248</td>\n",
       "      <td>2</td>\n",
       "      <td>{6: 41}</td>\n",
       "      <td>[6, 6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>simulated_trisomy_21</td>\n",
       "      <td>chr21_5016789</td>\n",
       "      <td>2</td>\n",
       "      <td>{9: 11, 10: 26}</td>\n",
       "      <td>[9, 10]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21493</th>\n",
       "      <td>simulated_trisomy_21</td>\n",
       "      <td>chr21_46688832</td>\n",
       "      <td>2</td>\n",
       "      <td>{12: 16, 16: 30}</td>\n",
       "      <td>[12, 16]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21494</th>\n",
       "      <td>simulated_trisomy_21</td>\n",
       "      <td>chr21_46689712</td>\n",
       "      <td>2</td>\n",
       "      <td>{3: 46}</td>\n",
       "      <td>[3, 3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21495</th>\n",
       "      <td>simulated_trisomy_21</td>\n",
       "      <td>chr21_46696282</td>\n",
       "      <td>2</td>\n",
       "      <td>{11: 26}</td>\n",
       "      <td>[11, 11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21496</th>\n",
       "      <td>simulated_trisomy_21</td>\n",
       "      <td>chr21_46697535</td>\n",
       "      <td>2</td>\n",
       "      <td>{20: 29}</td>\n",
       "      <td>[20, 20]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21497</th>\n",
       "      <td>simulated_trisomy_21</td>\n",
       "      <td>chr21_46699460</td>\n",
       "      <td>2</td>\n",
       "      <td>{3: 50}</td>\n",
       "      <td>[3, 3]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21498 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     sample          str_id  copy_number frequencies_gangstr  \\\n",
       "0      simulated_trisomy_21   chr21_5011212            2             {3: 56}   \n",
       "1      simulated_trisomy_21   chr21_5013077            2             {6: 50}   \n",
       "2      simulated_trisomy_21   chr21_5014892            2             {6: 44}   \n",
       "3      simulated_trisomy_21   chr21_5016248            2             {6: 41}   \n",
       "4      simulated_trisomy_21   chr21_5016789            2     {9: 11, 10: 26}   \n",
       "...                     ...             ...          ...                 ...   \n",
       "21493  simulated_trisomy_21  chr21_46688832            2    {12: 16, 16: 30}   \n",
       "21494  simulated_trisomy_21  chr21_46689712            2             {3: 46}   \n",
       "21495  simulated_trisomy_21  chr21_46696282            2            {11: 26}   \n",
       "21496  simulated_trisomy_21  chr21_46697535            2            {20: 29}   \n",
       "21497  simulated_trisomy_21  chr21_46699460            2             {3: 50}   \n",
       "\n",
       "      genotype_gangstr  \n",
       "0               [3, 3]  \n",
       "1               [6, 6]  \n",
       "2               [6, 6]  \n",
       "3               [6, 6]  \n",
       "4              [9, 10]  \n",
       "...                ...  \n",
       "21493         [12, 16]  \n",
       "21494           [3, 3]  \n",
       "21495         [11, 11]  \n",
       "21496         [20, 20]  \n",
       "21497           [3, 3]  \n",
       "\n",
       "[21498 rows x 5 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_repeats_gangstr, df_gangstr = file_io.dfs_from_vcf(filename=\"../../../data/simulated_reads/variants/GangSTR_simulated_trisomy_21.vcf\", samples=[\"simulated_trisomy_21\"], vcf_format=\"GangSTR\")\n",
    "df_gangstr = df_gangstr.assign(copy_number = 2)\n",
    "df_gangstr = df_gangstr.rename(columns = {\"frequencies\": \"frequencies_gangstr\", \"genotype\": \"genotype_gangstr\"})\n",
    "df_gangstr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "5a5afc43-f153-464a-b0a3-ec73c54e9136",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>str_id</th>\n",
       "      <th>copy_number</th>\n",
       "      <th>frequencies_hipstr</th>\n",
       "      <th>genotype_hipstr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tri21</td>\n",
       "      <td>chr21_5011212</td>\n",
       "      <td>2</td>\n",
       "      <td>{12: 58}</td>\n",
       "      <td>[12, 12]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tri21</td>\n",
       "      <td>chr21_5013077</td>\n",
       "      <td>2</td>\n",
       "      <td>{12: 54}</td>\n",
       "      <td>[12, 12]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tri21</td>\n",
       "      <td>chr21_5014892</td>\n",
       "      <td>2</td>\n",
       "      <td>{12: 46}</td>\n",
       "      <td>[12, 12]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tri21</td>\n",
       "      <td>chr21_5016248</td>\n",
       "      <td>2</td>\n",
       "      <td>{18: 41}</td>\n",
       "      <td>[18, 18]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tri21</td>\n",
       "      <td>chr21_5016789</td>\n",
       "      <td>2</td>\n",
       "      <td>{9: 12, 10: 33}</td>\n",
       "      <td>[9, 10]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20888</th>\n",
       "      <td>tri21</td>\n",
       "      <td>chr21_46688832</td>\n",
       "      <td>2</td>\n",
       "      <td>{24: 17, 32: 33}</td>\n",
       "      <td>[24, 32]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20889</th>\n",
       "      <td>tri21</td>\n",
       "      <td>chr21_46689712</td>\n",
       "      <td>2</td>\n",
       "      <td>{12: 47}</td>\n",
       "      <td>[12, 12]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20890</th>\n",
       "      <td>tri21</td>\n",
       "      <td>chr21_46696282</td>\n",
       "      <td>2</td>\n",
       "      <td>{11: 38}</td>\n",
       "      <td>[11, 11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20891</th>\n",
       "      <td>tri21</td>\n",
       "      <td>chr21_46697535</td>\n",
       "      <td>2</td>\n",
       "      <td>{20: 41}</td>\n",
       "      <td>[20, 20]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20892</th>\n",
       "      <td>tri21</td>\n",
       "      <td>chr21_46699460</td>\n",
       "      <td>2</td>\n",
       "      <td>{12: 50}</td>\n",
       "      <td>[12, 12]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20893 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample          str_id  copy_number frequencies_hipstr genotype_hipstr\n",
       "0      tri21   chr21_5011212            2           {12: 58}        [12, 12]\n",
       "1      tri21   chr21_5013077            2           {12: 54}        [12, 12]\n",
       "2      tri21   chr21_5014892            2           {12: 46}        [12, 12]\n",
       "3      tri21   chr21_5016248            2           {18: 41}        [18, 18]\n",
       "4      tri21   chr21_5016789            2    {9: 12, 10: 33}         [9, 10]\n",
       "...      ...             ...          ...                ...             ...\n",
       "20888  tri21  chr21_46688832            2   {24: 17, 32: 33}        [24, 32]\n",
       "20889  tri21  chr21_46689712            2           {12: 47}        [12, 12]\n",
       "20890  tri21  chr21_46696282            2           {11: 38}        [11, 11]\n",
       "20891  tri21  chr21_46697535            2           {20: 41}        [20, 20]\n",
       "20892  tri21  chr21_46699460            2           {12: 50}        [12, 12]\n",
       "\n",
       "[20893 rows x 5 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_repeats_hipstr, df_hipstr = file_io.dfs_from_vcf(filename=\"../../../data/simulated_reads/variants/HipSTR_simulated_trisomy_21.vcf\", samples=[\"tri21\"], vcf_format=\"HipSTR\")\n",
    "df_hipstr[\"copy_number\"] = 2\n",
    "df_hipstr = df_hipstr.rename(\n",
    "    columns = {\"frequencies\": \"frequencies_hipstr\", \"genotype\": \"genotype_hipstr\"}\n",
    ").assign(\n",
    "    genotype_hipstr = lambda x: [sorted(i) for i in x[\"genotype_hipstr\"]]\n",
    ")\n",
    "df_hipstr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "dab72e39-8b76-465c-b0d1-f3acbe41d199",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_id</th>\n",
       "      <th>frequencies_hipstr</th>\n",
       "      <th>genotype_hipstr</th>\n",
       "      <th>frequencies_gangstr</th>\n",
       "      <th>genotype_gangstr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr21_5011212</td>\n",
       "      <td>{12: 58}</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{3: 56}</td>\n",
       "      <td>[3, 3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr21_5013077</td>\n",
       "      <td>{12: 54}</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{6: 50}</td>\n",
       "      <td>[6, 6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr21_5014892</td>\n",
       "      <td>{12: 46}</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{6: 44}</td>\n",
       "      <td>[6, 6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr21_5016248</td>\n",
       "      <td>{18: 41}</td>\n",
       "      <td>[18, 18]</td>\n",
       "      <td>{6: 41}</td>\n",
       "      <td>[6, 6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr21_5016789</td>\n",
       "      <td>{9: 12, 10: 33}</td>\n",
       "      <td>[9, 10]</td>\n",
       "      <td>{9: 11, 10: 26}</td>\n",
       "      <td>[9, 10]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20872</th>\n",
       "      <td>chr21_46688832</td>\n",
       "      <td>{24: 17, 32: 33}</td>\n",
       "      <td>[24, 32]</td>\n",
       "      <td>{12: 16, 16: 30}</td>\n",
       "      <td>[12, 16]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20873</th>\n",
       "      <td>chr21_46689712</td>\n",
       "      <td>{12: 47}</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{3: 46}</td>\n",
       "      <td>[3, 3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20874</th>\n",
       "      <td>chr21_46696282</td>\n",
       "      <td>{11: 38}</td>\n",
       "      <td>[11, 11]</td>\n",
       "      <td>{11: 26}</td>\n",
       "      <td>[11, 11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20875</th>\n",
       "      <td>chr21_46697535</td>\n",
       "      <td>{20: 41}</td>\n",
       "      <td>[20, 20]</td>\n",
       "      <td>{20: 29}</td>\n",
       "      <td>[20, 20]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20876</th>\n",
       "      <td>chr21_46699460</td>\n",
       "      <td>{12: 50}</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{3: 50}</td>\n",
       "      <td>[3, 3]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20877 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               str_id frequencies_hipstr genotype_hipstr frequencies_gangstr  \\\n",
       "0       chr21_5011212           {12: 58}        [12, 12]             {3: 56}   \n",
       "1       chr21_5013077           {12: 54}        [12, 12]             {6: 50}   \n",
       "2       chr21_5014892           {12: 46}        [12, 12]             {6: 44}   \n",
       "3       chr21_5016248           {18: 41}        [18, 18]             {6: 41}   \n",
       "4       chr21_5016789    {9: 12, 10: 33}         [9, 10]     {9: 11, 10: 26}   \n",
       "...               ...                ...             ...                 ...   \n",
       "20872  chr21_46688832   {24: 17, 32: 33}        [24, 32]    {12: 16, 16: 30}   \n",
       "20873  chr21_46689712           {12: 47}        [12, 12]             {3: 46}   \n",
       "20874  chr21_46696282           {11: 38}        [11, 11]            {11: 26}   \n",
       "20875  chr21_46697535           {20: 41}        [20, 20]            {20: 29}   \n",
       "20876  chr21_46699460           {12: 50}        [12, 12]             {3: 50}   \n",
       "\n",
       "      genotype_gangstr  \n",
       "0               [3, 3]  \n",
       "1               [6, 6]  \n",
       "2               [6, 6]  \n",
       "3               [6, 6]  \n",
       "4              [9, 10]  \n",
       "...                ...  \n",
       "20872         [12, 16]  \n",
       "20873           [3, 3]  \n",
       "20874         [11, 11]  \n",
       "20875         [20, 20]  \n",
       "20876           [3, 3]  \n",
       "\n",
       "[20877 rows x 5 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_str_calls = (df_hipstr[[\"str_id\", \"frequencies_hipstr\", \"genotype_hipstr\"]]\n",
    "                   .merge(df_gangstr[[\"str_id\", \"frequencies_gangstr\", \"genotype_gangstr\"]], on=\"str_id\", how=\"inner\"))\n",
    "df_str_calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "a4d80116-412e-404c-973f-f2ae87db55ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_id</th>\n",
       "      <th>region_len_ref</th>\n",
       "      <th>region_len_mat</th>\n",
       "      <th>region_len_pat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr21_5354757</td>\n",
       "      <td>12</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr21_7205148</td>\n",
       "      <td>18</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr21_8987815</td>\n",
       "      <td>16</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr21_8987982</td>\n",
       "      <td>10</td>\n",
       "      <td>10.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr21_8988058</td>\n",
       "      <td>20</td>\n",
       "      <td>12.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18917</th>\n",
       "      <td>chr21_46688832</td>\n",
       "      <td>32</td>\n",
       "      <td>24.0</td>\n",
       "      <td>32.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18918</th>\n",
       "      <td>chr21_46689712</td>\n",
       "      <td>12</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18919</th>\n",
       "      <td>chr21_46696282</td>\n",
       "      <td>11</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18920</th>\n",
       "      <td>chr21_46697535</td>\n",
       "      <td>20</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18921</th>\n",
       "      <td>chr21_46699460</td>\n",
       "      <td>12</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18922 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               str_id  region_len_ref  region_len_mat  region_len_pat\n",
       "0       chr21_5354757              12            12.0            12.0\n",
       "1       chr21_7205148              18            18.0            18.0\n",
       "2       chr21_8987815              16            16.0            16.0\n",
       "3       chr21_8987982              10            10.0            10.0\n",
       "4       chr21_8988058              20            12.0            20.0\n",
       "...               ...             ...             ...             ...\n",
       "18917  chr21_46688832              32            24.0            32.0\n",
       "18918  chr21_46689712              12            12.0            12.0\n",
       "18919  chr21_46696282              11            11.0            11.0\n",
       "18920  chr21_46697535              20            20.0            20.0\n",
       "18921  chr21_46699460              12            12.0            12.0\n",
       "\n",
       "[18922 rows x 4 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_repeats_hg002 = pd.read_csv(\"../../../data/HG002/variants/HG002_GRCh38_STR_lengths.csv\")\n",
    "df_repeats_hg002 = df_repeats_hg002[df_repeats_hg002[\"str_id\"].isin(df_str_calls[\"str_id\"])].reset_index(drop=True)\n",
    "df_repeats_hg002"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "aca6621e-93be-4d65-90b6-6c35cc3646ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>period</th>\n",
       "      <th>unit</th>\n",
       "      <th>str_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50287</th>\n",
       "      <td>chr21</td>\n",
       "      <td>5011211</td>\n",
       "      <td>5011223</td>\n",
       "      <td>4</td>\n",
       "      <td>AATC</td>\n",
       "      <td>chr21_5011212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50288</th>\n",
       "      <td>chr21</td>\n",
       "      <td>5013076</td>\n",
       "      <td>5013088</td>\n",
       "      <td>2</td>\n",
       "      <td>CA</td>\n",
       "      <td>chr21_5013077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50289</th>\n",
       "      <td>chr21</td>\n",
       "      <td>5014891</td>\n",
       "      <td>5014903</td>\n",
       "      <td>2</td>\n",
       "      <td>CA</td>\n",
       "      <td>chr21_5014892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50290</th>\n",
       "      <td>chr21</td>\n",
       "      <td>5016247</td>\n",
       "      <td>5016265</td>\n",
       "      <td>3</td>\n",
       "      <td>GAG</td>\n",
       "      <td>chr21_5016248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50291</th>\n",
       "      <td>chr21</td>\n",
       "      <td>5016788</td>\n",
       "      <td>5016798</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>chr21_5016789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53380</th>\n",
       "      <td>chr21</td>\n",
       "      <td>46688831</td>\n",
       "      <td>46688863</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>chr21_46688832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53381</th>\n",
       "      <td>chr21</td>\n",
       "      <td>46689711</td>\n",
       "      <td>46689723</td>\n",
       "      <td>4</td>\n",
       "      <td>ATTT</td>\n",
       "      <td>chr21_46689712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53382</th>\n",
       "      <td>chr21</td>\n",
       "      <td>46696281</td>\n",
       "      <td>46696292</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>chr21_46696282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53383</th>\n",
       "      <td>chr21</td>\n",
       "      <td>46697534</td>\n",
       "      <td>46697554</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>chr21_46697535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53384</th>\n",
       "      <td>chr21</td>\n",
       "      <td>46699459</td>\n",
       "      <td>46699471</td>\n",
       "      <td>4</td>\n",
       "      <td>GGGC</td>\n",
       "      <td>chr21_46699460</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3098 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr     start       end  period  unit          str_id\n",
       "50287  chr21   5011211   5011223       4  AATC   chr21_5011212\n",
       "50288  chr21   5013076   5013088       2    CA   chr21_5013077\n",
       "50289  chr21   5014891   5014903       2    CA   chr21_5014892\n",
       "50290  chr21   5016247   5016265       3   GAG   chr21_5016248\n",
       "50291  chr21   5016788   5016798       1     G   chr21_5016789\n",
       "...      ...       ...       ...     ...   ...             ...\n",
       "53380  chr21  46688831  46688863       2    AC  chr21_46688832\n",
       "53381  chr21  46689711  46689723       4  ATTT  chr21_46689712\n",
       "53382  chr21  46696281  46696292       1     T  chr21_46696282\n",
       "53383  chr21  46697534  46697554       1     A  chr21_46697535\n",
       "53384  chr21  46699459  46699471       4  GGGC  chr21_46699460\n",
       "\n",
       "[3098 rows x 6 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "names = [\"chr\", \"start\", \"end\", \"period\", \"unit\"]\n",
    "df_strs_in_segdup = (pd.read_csv(\"../../../data/hg38_ver13_0boe_mononucleotides_in_segdup.bed\", sep=\"\\t\", names=names)\n",
    "                        .query(\"chr == 'chr21'\")\n",
    "                        .assign(str_id = lambda x: [f\"{i}_{j}\" for i, j in zip(x[\"chr\"], x[\"start\"] + 1)]))\n",
    "df_strs_in_segdup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "88553708-f479-4156-b154-aa16cc19fd23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_id</th>\n",
       "      <th>period</th>\n",
       "      <th>ref</th>\n",
       "      <th>genotype_gangstr</th>\n",
       "      <th>frequencies_gangstr</th>\n",
       "      <th>genotype_hipstr</th>\n",
       "      <th>frequencies_hipstr</th>\n",
       "      <th>genotype_sim</th>\n",
       "      <th>n_distinct_alleles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr21_10324665</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>[15, 15]</td>\n",
       "      <td>{3: 17}</td>\n",
       "      <td>[15, 15]</td>\n",
       "      <td>{15: 23}</td>\n",
       "      <td>[15, 15, 15]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr21_12974501</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>[10, 10]</td>\n",
       "      <td>{10: 35}</td>\n",
       "      <td>[10, 10]</td>\n",
       "      <td>{10: 53}</td>\n",
       "      <td>[10, 10, 10]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr21_13987256</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{2: 37}</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{12: 48}</td>\n",
       "      <td>[12, 12, 12]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr21_13987386</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>[10, 10]</td>\n",
       "      <td>{2: 37}</td>\n",
       "      <td>[10, 10]</td>\n",
       "      <td>{10: 54}</td>\n",
       "      <td>[10, 10, 10]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr21_13998484</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>[12, 13]</td>\n",
       "      <td>{12: 18, 13: 14, 14: 9}</td>\n",
       "      <td>[12, 13]</td>\n",
       "      <td>{-8: 1, 12: 24, 13: 22, 14: 15}</td>\n",
       "      <td>[12, 13, 14]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17743</th>\n",
       "      <td>chr21_46657138</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>[10, 10]</td>\n",
       "      <td>{10: 45}</td>\n",
       "      <td>[10, 10]</td>\n",
       "      <td>{10: 53}</td>\n",
       "      <td>[10, 10, 10]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17744</th>\n",
       "      <td>chr21_46657377</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>[14, 14]</td>\n",
       "      <td>{14: 19}</td>\n",
       "      <td>[14, 14]</td>\n",
       "      <td>{14: 29}</td>\n",
       "      <td>[14, 14, 14]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17745</th>\n",
       "      <td>chr21_46666129</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>[11, 11]</td>\n",
       "      <td>{11: 28}</td>\n",
       "      <td>[11, 11]</td>\n",
       "      <td>{11: 37}</td>\n",
       "      <td>[11, 11, 11]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17746</th>\n",
       "      <td>chr21_46668752</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>[12, 14]</td>\n",
       "      <td>{12: 20, 14: 22}</td>\n",
       "      <td>[12, 14]</td>\n",
       "      <td>{12: 24, 14: 29}</td>\n",
       "      <td>[12, 14, 14]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17747</th>\n",
       "      <td>chr21_46672634</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{12: 57}</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{12: 71}</td>\n",
       "      <td>[12, 12, 12]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17748 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               str_id  period  ref genotype_gangstr      frequencies_gangstr  \\\n",
       "0      chr21_10324665       5    3         [15, 15]                  {3: 17}   \n",
       "1      chr21_12974501       1   10         [10, 10]                 {10: 35}   \n",
       "2      chr21_13987256       6    2         [12, 12]                  {2: 37}   \n",
       "3      chr21_13987386       5    2         [10, 10]                  {2: 37}   \n",
       "4      chr21_13998484       1   12         [12, 13]  {12: 18, 13: 14, 14: 9}   \n",
       "...               ...     ...  ...              ...                      ...   \n",
       "17743  chr21_46657138       1   10         [10, 10]                 {10: 45}   \n",
       "17744  chr21_46657377       1   14         [14, 14]                 {14: 19}   \n",
       "17745  chr21_46666129       1   11         [11, 11]                 {11: 28}   \n",
       "17746  chr21_46668752       1   14         [12, 14]         {12: 20, 14: 22}   \n",
       "17747  chr21_46672634       1   12         [12, 12]                 {12: 57}   \n",
       "\n",
       "      genotype_hipstr               frequencies_hipstr  genotype_sim  \\\n",
       "0            [15, 15]                         {15: 23}  [15, 15, 15]   \n",
       "1            [10, 10]                         {10: 53}  [10, 10, 10]   \n",
       "2            [12, 12]                         {12: 48}  [12, 12, 12]   \n",
       "3            [10, 10]                         {10: 54}  [10, 10, 10]   \n",
       "4            [12, 13]  {-8: 1, 12: 24, 13: 22, 14: 15}  [12, 13, 14]   \n",
       "...               ...                              ...           ...   \n",
       "17743        [10, 10]                         {10: 53}  [10, 10, 10]   \n",
       "17744        [14, 14]                         {14: 29}  [14, 14, 14]   \n",
       "17745        [11, 11]                         {11: 37}  [11, 11, 11]   \n",
       "17746        [12, 14]                 {12: 24, 14: 29}  [12, 14, 14]   \n",
       "17747        [12, 12]                         {12: 71}  [12, 12, 12]   \n",
       "\n",
       "       n_distinct_alleles  \n",
       "0                       1  \n",
       "1                       1  \n",
       "2                       1  \n",
       "3                       1  \n",
       "4                       3  \n",
       "...                   ...  \n",
       "17743                   1  \n",
       "17744                   1  \n",
       "17745                   1  \n",
       "17746                   2  \n",
       "17747                   1  \n",
       "\n",
       "[17748 rows x 9 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_comparison = (df_repeats_hg002\n",
    "     .merge(df_str_calls, on=\"str_id\", how=\"inner\")\n",
    "     .merge(df_repeats_gangstr[[\"str_id\", \"period\", \"ref\"]], how=\"inner\")\n",
    "     .assign(region_len_mat = lambda x: x[\"region_len_mat\"].convert_dtypes(int),\n",
    "             region_len_pat = lambda x: x[\"region_len_pat\"].convert_dtypes(int),\n",
    "             genotype_gangstr = lambda x: x[\"genotype_gangstr\"] * x[\"period\"],\n",
    "             genotype_sim = lambda x: [sorted([i, j, k]) for i, j, k in zip(x[\"region_len_mat\"], x[\"region_len_pat\"], x[\"region_len_ref\"])],\n",
    "             n_distinct_alleles = lambda x: [len(set(i)) for i in x[\"genotype_sim\"]])\n",
    "     .filter([\"str_id\", \"period\", \"ref\", \"unit\", \"genotype_gangstr\", \"frequencies_gangstr\", \"genotype_hipstr\", \"frequencies_hipstr\", \"genotype_sim\", \"n_distinct_alleles\"], axis=1))\n",
    "\n",
    "df_comparison = df_comparison[~df_comparison[\"str_id\"].isin(df_strs_in_segdup[\"str_id\"])].reset_index(drop=True)\n",
    "df_comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "4c0f292c-f746-41b6-9c11-b23b96735f5a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_distinct_alleles\n",
      "1    13113\n",
      "2     3825\n",
      "3      810\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "n_distinct_alleles\n",
       "1    0.738844\n",
       "2    0.215517\n",
       "3    0.045639\n",
       "Name: proportion, dtype: float64"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(df_comparison[\"n_distinct_alleles\"].value_counts())\n",
    "df_comparison[\"n_distinct_alleles\"].value_counts(normalize=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "71f924d7-6e2e-492d-83cb-a6ee13138c1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_id</th>\n",
       "      <th>period</th>\n",
       "      <th>ref</th>\n",
       "      <th>genotype_gangstr</th>\n",
       "      <th>frequencies_gangstr</th>\n",
       "      <th>genotype_hipstr</th>\n",
       "      <th>frequencies_hipstr</th>\n",
       "      <th>genotype_sim</th>\n",
       "      <th>n_distinct_alleles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr21_14094421</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>[9, 10]</td>\n",
       "      <td>{9: 32, 10: 9}</td>\n",
       "      <td>[9, 10]</td>\n",
       "      <td>{9: 43, 10: 16}</td>\n",
       "      <td>[9, 9, 10]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr21_14114178</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{15: 7, 16: 27}</td>\n",
       "      <td>[15, 16]</td>\n",
       "      <td>{15: 10, 16: 34}</td>\n",
       "      <td>[15, 16, 16]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr21_14124995</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>[10, 11]</td>\n",
       "      <td>{10: 18, 11: 32}</td>\n",
       "      <td>[10, 11]</td>\n",
       "      <td>{10: 22, 11: 42}</td>\n",
       "      <td>[10, 11, 11]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>chr21_14141387</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>[11, 13]</td>\n",
       "      <td>{11: 26, 13: 10}</td>\n",
       "      <td>[11, 13]</td>\n",
       "      <td>{11: 35, 13: 16}</td>\n",
       "      <td>[11, 11, 13]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr21_14163093</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>[14, 16]</td>\n",
       "      <td>{7: 33, 8: 15}</td>\n",
       "      <td>[14, 16]</td>\n",
       "      <td>{14: 38, 16: 20}</td>\n",
       "      <td>[14, 14, 16]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            str_id  period  ref genotype_gangstr frequencies_gangstr  \\\n",
       "21  chr21_14094421       1   10          [9, 10]      {9: 32, 10: 9}   \n",
       "27  chr21_14114178       1   16         [16, 16]     {15: 7, 16: 27}   \n",
       "28  chr21_14124995       1   11         [10, 11]    {10: 18, 11: 32}   \n",
       "37  chr21_14141387       1   13         [11, 13]    {11: 26, 13: 10}   \n",
       "45  chr21_14163093       2    8         [14, 16]      {7: 33, 8: 15}   \n",
       "\n",
       "   genotype_hipstr frequencies_hipstr  genotype_sim  n_distinct_alleles  \n",
       "21         [9, 10]    {9: 43, 10: 16}    [9, 9, 10]                   2  \n",
       "27        [15, 16]   {15: 10, 16: 34}  [15, 16, 16]                   2  \n",
       "28        [10, 11]   {10: 22, 11: 42}  [10, 11, 11]                   2  \n",
       "37        [11, 13]   {11: 35, 13: 16}  [11, 11, 13]                   2  \n",
       "45        [14, 16]   {14: 38, 16: 20}  [14, 14, 16]                   2  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_comparison.query(\"n_distinct_alleles == 2\").head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "5e17b217-f558-42bb-b80a-9f3bc1090197",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_distinct_alleles</th>\n",
       "      <th>distinct_gangstr</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.999542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.924706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.075294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.983951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.016049</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   n_distinct_alleles  distinct_gangstr  proportion\n",
       "0                   1                 1    0.999542\n",
       "1                   1                 2    0.000458\n",
       "2                   2                 2    0.924706\n",
       "3                   2                 1    0.075294\n",
       "4                   3                 2    0.983951\n",
       "5                   3                 1    0.016049"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_comparison.dropna(subset=\"frequencies_gangstr\").assign(\n",
    "    distinct_gangstr = lambda x: [len(set(i)) for i in x[\"genotype_gangstr\"]],\n",
    "    missing_gangstr = lambda x: x[\"n_distinct_alleles\"] - x[\"distinct_gangstr\"],\n",
    ").filter([\"n_distinct_alleles\", \"distinct_gangstr\"]).groupby(\"n_distinct_alleles\", as_index=False).value_counts(normalize=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "4ef407c4-ecb8-422a-9192-696ac5a19208",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_id</th>\n",
       "      <th>period</th>\n",
       "      <th>ref</th>\n",
       "      <th>genotype_gangstr</th>\n",
       "      <th>frequencies_gangstr</th>\n",
       "      <th>genotype_hipstr</th>\n",
       "      <th>frequencies_hipstr</th>\n",
       "      <th>genotype_sim</th>\n",
       "      <th>n_distinct_alleles</th>\n",
       "      <th>distinct_gangstr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr21_14114178</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{15: 7, 16: 27}</td>\n",
       "      <td>[15, 16]</td>\n",
       "      <td>{15: 10, 16: 34}</td>\n",
       "      <td>[15, 16, 16]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>chr21_14191377</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>[17, 17]</td>\n",
       "      <td>{17: 29, 18: 7}</td>\n",
       "      <td>[17, 18]</td>\n",
       "      <td>{17: 37, 18: 13}</td>\n",
       "      <td>[17, 17, 18]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>chr21_14235352</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{8: 56}</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{16: 57}</td>\n",
       "      <td>[16, 16, 17]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>chr21_14243744</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>[8, 8]</td>\n",
       "      <td>{2: 47}</td>\n",
       "      <td>[8, 8]</td>\n",
       "      <td>{8: 58}</td>\n",
       "      <td>[6, 8, 8]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>chr21_14268035</td>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>[24, 24]</td>\n",
       "      <td>{12: 25}</td>\n",
       "      <td>[24, 24]</td>\n",
       "      <td>{24: 43}</td>\n",
       "      <td>[21, 24, 24]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>chr21_14303570</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>[40, 40]</td>\n",
       "      <td>{8: 46}</td>\n",
       "      <td>[40, 40]</td>\n",
       "      <td>{40: 46}</td>\n",
       "      <td>[40, 40, 42]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>chr21_14433765</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{4: 42}</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{16: 43}</td>\n",
       "      <td>[16, 17, 17]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>chr21_14554699</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>[19, 19]</td>\n",
       "      <td>{18: 7, 19: 28}</td>\n",
       "      <td>[18, 19]</td>\n",
       "      <td>{1: 1, 13: 1, 18: 13, 19: 37}</td>\n",
       "      <td>[18, 19, 19]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>chr21_14760579</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{3: 48}</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{12: 50}</td>\n",
       "      <td>[11, 11, 12]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>chr21_14903033</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{4: 44}</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{12: 44}</td>\n",
       "      <td>[4, 4, 12]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             str_id  period  ref genotype_gangstr frequencies_gangstr  \\\n",
       "27   chr21_14114178       1   16         [16, 16]     {15: 7, 16: 27}   \n",
       "55   chr21_14191377       1   17         [17, 17]     {17: 29, 18: 7}   \n",
       "75   chr21_14235352       2    8         [16, 16]             {8: 56}   \n",
       "78   chr21_14243744       4    2           [8, 8]             {2: 47}   \n",
       "88   chr21_14268035       2   12         [24, 24]            {12: 25}   \n",
       "107  chr21_14303570       5    8         [40, 40]             {8: 46}   \n",
       "170  chr21_14433765       4    4         [16, 16]             {4: 42}   \n",
       "241  chr21_14554699       1   18         [19, 19]     {18: 7, 19: 28}   \n",
       "341  chr21_14760579       4    3         [12, 12]             {3: 48}   \n",
       "405  chr21_14903033       3    4         [12, 12]             {4: 44}   \n",
       "\n",
       "    genotype_hipstr             frequencies_hipstr  genotype_sim  \\\n",
       "27         [15, 16]               {15: 10, 16: 34}  [15, 16, 16]   \n",
       "55         [17, 18]               {17: 37, 18: 13}  [17, 17, 18]   \n",
       "75         [16, 16]                       {16: 57}  [16, 16, 17]   \n",
       "78           [8, 8]                        {8: 58}     [6, 8, 8]   \n",
       "88         [24, 24]                       {24: 43}  [21, 24, 24]   \n",
       "107        [40, 40]                       {40: 46}  [40, 40, 42]   \n",
       "170        [16, 16]                       {16: 43}  [16, 17, 17]   \n",
       "241        [18, 19]  {1: 1, 13: 1, 18: 13, 19: 37}  [18, 19, 19]   \n",
       "341        [12, 12]                       {12: 50}  [11, 11, 12]   \n",
       "405        [12, 12]                       {12: 44}    [4, 4, 12]   \n",
       "\n",
       "     n_distinct_alleles  distinct_gangstr  \n",
       "27                    2                 1  \n",
       "55                    2                 1  \n",
       "75                    2                 1  \n",
       "78                    2                 1  \n",
       "88                    2                 1  \n",
       "107                   2                 1  \n",
       "170                   2                 1  \n",
       "241                   2                 1  \n",
       "341                   2                 1  \n",
       "405                   2                 1  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_comparison.query(\"n_distinct_alleles == 2\").assign(\n",
    "    distinct_gangstr = lambda x: [len(set(i)) for i in x[\"genotype_gangstr\"]]\n",
    ").query(\"distinct_gangstr == 1\").head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "db9fd5a8-95a3-42b4-bdc9-4b70a22ac549",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_id</th>\n",
       "      <th>period</th>\n",
       "      <th>ref</th>\n",
       "      <th>genotype_gangstr</th>\n",
       "      <th>frequencies_gangstr</th>\n",
       "      <th>genotype_hipstr</th>\n",
       "      <th>frequencies_hipstr</th>\n",
       "      <th>genotype_sim</th>\n",
       "      <th>n_distinct_alleles</th>\n",
       "      <th>distinct_gangstr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>937</th>\n",
       "      <td>chr21_15994635</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>[20, 20]</td>\n",
       "      <td>{5: 45}</td>\n",
       "      <td>[20, 20]</td>\n",
       "      <td>{20: 47}</td>\n",
       "      <td>[20, 33, 37]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2768</th>\n",
       "      <td>chr21_19488881</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>[44, 44]</td>\n",
       "      <td>{22: 32}</td>\n",
       "      <td>[44, 44]</td>\n",
       "      <td>{44: 35}</td>\n",
       "      <td>[35, 37, 44]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3540</th>\n",
       "      <td>chr21_20982238</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>[22, 22]</td>\n",
       "      <td>{11: 32}</td>\n",
       "      <td>[22, 22]</td>\n",
       "      <td>{22: 45}</td>\n",
       "      <td>[13, 22, 25]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5994</th>\n",
       "      <td>chr21_25572109</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>[22, 22]</td>\n",
       "      <td>{11: 36}</td>\n",
       "      <td>[22, 22]</td>\n",
       "      <td>{22: 50}</td>\n",
       "      <td>[22, 23, 25]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7300</th>\n",
       "      <td>chr21_28173037</td>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>[24, 24]</td>\n",
       "      <td>{12: 27, 13: 7}</td>\n",
       "      <td>[24, 24]</td>\n",
       "      <td>{24: 42, 26: 10}</td>\n",
       "      <td>[24, 26, 27]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11671</th>\n",
       "      <td>chr21_35446085</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>[30, 30]</td>\n",
       "      <td>{15: 26}</td>\n",
       "      <td>[30, 30]</td>\n",
       "      <td>{30: 40}</td>\n",
       "      <td>[27, 29, 30]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12415</th>\n",
       "      <td>chr21_36462935</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>[8, 8]</td>\n",
       "      <td>{2: 47}</td>\n",
       "      <td>[8, 8]</td>\n",
       "      <td>{8: 59}</td>\n",
       "      <td>[6, 7, 8]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13182</th>\n",
       "      <td>chr21_37597069</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{4: 47}</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{16: 49}</td>\n",
       "      <td>[16, 19, 21]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14097</th>\n",
       "      <td>chr21_39337023</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>[15, 15]</td>\n",
       "      <td>{3: 39}</td>\n",
       "      <td>[15, 15]</td>\n",
       "      <td>{15: 39}</td>\n",
       "      <td>[15, 16, 22]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14530</th>\n",
       "      <td>chr21_40106894</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>[10, 10]</td>\n",
       "      <td>{2: 44}</td>\n",
       "      <td>[10, 10]</td>\n",
       "      <td>{10: 50}</td>\n",
       "      <td>[9, 10, 12]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15425</th>\n",
       "      <td>chr21_41776804</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{4: 45}</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{16: 47}</td>\n",
       "      <td>[16, 17, 18]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17359</th>\n",
       "      <td>chr21_45909791</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>[10, 10]</td>\n",
       "      <td>{10: 19}</td>\n",
       "      <td>[8, 10]</td>\n",
       "      <td>{-10: 1, -6: 2, -4: 1, 1: 7, 2: 2, 10: 29}</td>\n",
       "      <td>[10, 180, 355]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17735</th>\n",
       "      <td>chr21_46645267</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>[14, 14]</td>\n",
       "      <td>{14: 8}</td>\n",
       "      <td>[14]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[14, 119, 120]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               str_id  period  ref genotype_gangstr frequencies_gangstr  \\\n",
       "937    chr21_15994635       4    5         [20, 20]             {5: 45}   \n",
       "2768   chr21_19488881       2   22         [44, 44]            {22: 32}   \n",
       "3540   chr21_20982238       2   11         [22, 22]            {11: 32}   \n",
       "5994   chr21_25572109       2   11         [22, 22]            {11: 36}   \n",
       "7300   chr21_28173037       2   12         [24, 24]     {12: 27, 13: 7}   \n",
       "11671  chr21_35446085       2   15         [30, 30]            {15: 26}   \n",
       "12415  chr21_36462935       4    2           [8, 8]             {2: 47}   \n",
       "13182  chr21_37597069       4    4         [16, 16]             {4: 47}   \n",
       "14097  chr21_39337023       5    3         [15, 15]             {3: 39}   \n",
       "14530  chr21_40106894       5    2         [10, 10]             {2: 44}   \n",
       "15425  chr21_41776804       4    4         [16, 16]             {4: 45}   \n",
       "17359  chr21_45909791       1   10         [10, 10]            {10: 19}   \n",
       "17735  chr21_46645267       1   14         [14, 14]             {14: 8}   \n",
       "\n",
       "      genotype_hipstr                          frequencies_hipstr  \\\n",
       "937          [20, 20]                                    {20: 47}   \n",
       "2768         [44, 44]                                    {44: 35}   \n",
       "3540         [22, 22]                                    {22: 45}   \n",
       "5994         [22, 22]                                    {22: 50}   \n",
       "7300         [24, 24]                            {24: 42, 26: 10}   \n",
       "11671        [30, 30]                                    {30: 40}   \n",
       "12415          [8, 8]                                     {8: 59}   \n",
       "13182        [16, 16]                                    {16: 49}   \n",
       "14097        [15, 15]                                    {15: 39}   \n",
       "14530        [10, 10]                                    {10: 50}   \n",
       "15425        [16, 16]                                    {16: 47}   \n",
       "17359         [8, 10]  {-10: 1, -6: 2, -4: 1, 1: 7, 2: 2, 10: 29}   \n",
       "17735            [14]                                         NaN   \n",
       "\n",
       "         genotype_sim  n_distinct_alleles  distinct_gangstr  \n",
       "937      [20, 33, 37]                   3                 1  \n",
       "2768     [35, 37, 44]                   3                 1  \n",
       "3540     [13, 22, 25]                   3                 1  \n",
       "5994     [22, 23, 25]                   3                 1  \n",
       "7300     [24, 26, 27]                   3                 1  \n",
       "11671    [27, 29, 30]                   3                 1  \n",
       "12415       [6, 7, 8]                   3                 1  \n",
       "13182    [16, 19, 21]                   3                 1  \n",
       "14097    [15, 16, 22]                   3                 1  \n",
       "14530     [9, 10, 12]                   3                 1  \n",
       "15425    [16, 17, 18]                   3                 1  \n",
       "17359  [10, 180, 355]                   3                 1  \n",
       "17735  [14, 119, 120]                   3                 1  "
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# chr21_14213292\n",
    "df_comparison.query(\"n_distinct_alleles == 3\").assign(\n",
    "    distinct_gangstr = lambda x: [len(set(i)) for i in x[\"genotype_gangstr\"]]\n",
    ").query(\"distinct_gangstr == 1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "f964e55f-a638-4d5f-95c1-fa75b1372572",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_distinct_alleles</th>\n",
       "      <th>missing_hipstr</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>13113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   n_distinct_alleles  missing_hipstr  count\n",
       "0                   1               0  13113\n",
       "1                   2               0   3575\n",
       "2                   2               1    248\n",
       "3                   3               1    783\n",
       "4                   3               2     11"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_comparison.dropna(subset=\"frequencies_hipstr\").assign(\n",
    "    distinct_hipstr = lambda x: [len(set(i)) for i in x[\"genotype_hipstr\"]],\n",
    "    missing_hipstr = lambda x: x[\"n_distinct_alleles\"] - x[\"distinct_hipstr\"],\n",
    ").filter([\"n_distinct_alleles\", \"missing_hipstr\"]).groupby(\"n_distinct_alleles\", as_index=False).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "a3037df2-bc92-4d55-8337-6f5362147c35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_distinct_alleles</th>\n",
       "      <th>distinct_hipstr</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.935129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.064871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.986146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.013854</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   n_distinct_alleles  distinct_hipstr  proportion\n",
       "0                   1                1    1.000000\n",
       "1                   2                2    0.935129\n",
       "2                   2                1    0.064871\n",
       "3                   3                2    0.986146\n",
       "4                   3                1    0.013854"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_comparison.dropna(subset=\"frequencies_hipstr\").assign(\n",
    "    distinct_hipstr = lambda x: [len(set(i)) for i in x[\"genotype_hipstr\"]],\n",
    "    missing_hipstr = lambda x: x[\"n_distinct_alleles\"] - x[\"distinct_hipstr\"],\n",
    ").filter([\"n_distinct_alleles\", \"distinct_hipstr\"]).groupby(\"n_distinct_alleles\", as_index=False).value_counts(normalize=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "88705491-9876-4eef-80fd-4c44d540aba9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_id</th>\n",
       "      <th>period</th>\n",
       "      <th>ref</th>\n",
       "      <th>genotype_gangstr</th>\n",
       "      <th>frequencies_gangstr</th>\n",
       "      <th>genotype_hipstr</th>\n",
       "      <th>frequencies_hipstr</th>\n",
       "      <th>genotype_sim</th>\n",
       "      <th>n_distinct_alleles</th>\n",
       "      <th>distinct_hipstr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>chr21_14235352</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{8: 56}</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{16: 57}</td>\n",
       "      <td>[16, 16, 17]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>chr21_14243744</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>[8, 8]</td>\n",
       "      <td>{2: 47}</td>\n",
       "      <td>[8, 8]</td>\n",
       "      <td>{8: 58}</td>\n",
       "      <td>[6, 8, 8]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>chr21_14268035</td>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>[24, 24]</td>\n",
       "      <td>{12: 25}</td>\n",
       "      <td>[24, 24]</td>\n",
       "      <td>{24: 43}</td>\n",
       "      <td>[21, 24, 24]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>chr21_14303570</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>[40, 40]</td>\n",
       "      <td>{8: 46}</td>\n",
       "      <td>[40, 40]</td>\n",
       "      <td>{40: 46}</td>\n",
       "      <td>[40, 40, 42]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>chr21_14433765</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{4: 42}</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{16: 43}</td>\n",
       "      <td>[16, 17, 17]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>chr21_14534525</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>[11, 12]</td>\n",
       "      <td>{11: 9, 12: 23}</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{-4: 1, 11: 13, 12: 36}</td>\n",
       "      <td>[11, 12, 12]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>chr21_14740792</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>[10, 11]</td>\n",
       "      <td>{10: 31, 11: 9}</td>\n",
       "      <td>[10, 10]</td>\n",
       "      <td>{10: 39, 11: 11}</td>\n",
       "      <td>[10, 10, 11]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>chr21_14760579</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{3: 48}</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{12: 50}</td>\n",
       "      <td>[11, 11, 12]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>chr21_14903033</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{4: 44}</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>{12: 44}</td>\n",
       "      <td>[4, 4, 12]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>chr21_14909188</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>[24, 25]</td>\n",
       "      <td>{24: 6, 25: 17}</td>\n",
       "      <td>[25, 25]</td>\n",
       "      <td>{22: 1, 24: 7, 25: 24}</td>\n",
       "      <td>[24, 25, 25]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             str_id  period  ref genotype_gangstr frequencies_gangstr  \\\n",
       "75   chr21_14235352       2    8         [16, 16]             {8: 56}   \n",
       "78   chr21_14243744       4    2           [8, 8]             {2: 47}   \n",
       "88   chr21_14268035       2   12         [24, 24]            {12: 25}   \n",
       "107  chr21_14303570       5    8         [40, 40]             {8: 46}   \n",
       "170  chr21_14433765       4    4         [16, 16]             {4: 42}   \n",
       "225  chr21_14534525       1   11         [11, 12]     {11: 9, 12: 23}   \n",
       "334  chr21_14740792       1   11         [10, 11]     {10: 31, 11: 9}   \n",
       "341  chr21_14760579       4    3         [12, 12]             {3: 48}   \n",
       "405  chr21_14903033       3    4         [12, 12]             {4: 44}   \n",
       "409  chr21_14909188       1   24         [24, 25]     {24: 6, 25: 17}   \n",
       "\n",
       "    genotype_hipstr       frequencies_hipstr  genotype_sim  \\\n",
       "75         [16, 16]                 {16: 57}  [16, 16, 17]   \n",
       "78           [8, 8]                  {8: 58}     [6, 8, 8]   \n",
       "88         [24, 24]                 {24: 43}  [21, 24, 24]   \n",
       "107        [40, 40]                 {40: 46}  [40, 40, 42]   \n",
       "170        [16, 16]                 {16: 43}  [16, 17, 17]   \n",
       "225        [12, 12]  {-4: 1, 11: 13, 12: 36}  [11, 12, 12]   \n",
       "334        [10, 10]         {10: 39, 11: 11}  [10, 10, 11]   \n",
       "341        [12, 12]                 {12: 50}  [11, 11, 12]   \n",
       "405        [12, 12]                 {12: 44}    [4, 4, 12]   \n",
       "409        [25, 25]   {22: 1, 24: 7, 25: 24}  [24, 25, 25]   \n",
       "\n",
       "     n_distinct_alleles  distinct_hipstr  \n",
       "75                    2                1  \n",
       "78                    2                1  \n",
       "88                    2                1  \n",
       "107                   2                1  \n",
       "170                   2                1  \n",
       "225                   2                1  \n",
       "334                   2                1  \n",
       "341                   2                1  \n",
       "405                   2                1  \n",
       "409                   2                1  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_comparison.dropna(subset=\"frequencies_hipstr\").query(\"n_distinct_alleles == 2\").assign(\n",
    "    distinct_hipstr = lambda x: [len(set(i)) for i in x[\"genotype_hipstr\"]]\n",
    ").query(\"distinct_hipstr == 1\").head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "1a78ab55-e41f-462a-b36a-7e9d1f56c049",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_id</th>\n",
       "      <th>period</th>\n",
       "      <th>ref</th>\n",
       "      <th>genotype_gangstr</th>\n",
       "      <th>frequencies_gangstr</th>\n",
       "      <th>genotype_hipstr</th>\n",
       "      <th>frequencies_hipstr</th>\n",
       "      <th>genotype_sim</th>\n",
       "      <th>n_distinct_alleles</th>\n",
       "      <th>distinct_hipstr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>937</th>\n",
       "      <td>chr21_15994635</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>[20, 20]</td>\n",
       "      <td>{5: 45}</td>\n",
       "      <td>[20, 20]</td>\n",
       "      <td>{20: 47}</td>\n",
       "      <td>[20, 33, 37]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2768</th>\n",
       "      <td>chr21_19488881</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>[44, 44]</td>\n",
       "      <td>{22: 32}</td>\n",
       "      <td>[44, 44]</td>\n",
       "      <td>{44: 35}</td>\n",
       "      <td>[35, 37, 44]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3540</th>\n",
       "      <td>chr21_20982238</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>[22, 22]</td>\n",
       "      <td>{11: 32}</td>\n",
       "      <td>[22, 22]</td>\n",
       "      <td>{22: 45}</td>\n",
       "      <td>[13, 22, 25]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5994</th>\n",
       "      <td>chr21_25572109</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>[22, 22]</td>\n",
       "      <td>{11: 36}</td>\n",
       "      <td>[22, 22]</td>\n",
       "      <td>{22: 50}</td>\n",
       "      <td>[22, 23, 25]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7300</th>\n",
       "      <td>chr21_28173037</td>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>[24, 24]</td>\n",
       "      <td>{12: 27, 13: 7}</td>\n",
       "      <td>[24, 24]</td>\n",
       "      <td>{24: 42, 26: 10}</td>\n",
       "      <td>[24, 26, 27]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11671</th>\n",
       "      <td>chr21_35446085</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>[30, 30]</td>\n",
       "      <td>{15: 26}</td>\n",
       "      <td>[30, 30]</td>\n",
       "      <td>{30: 40}</td>\n",
       "      <td>[27, 29, 30]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12415</th>\n",
       "      <td>chr21_36462935</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>[8, 8]</td>\n",
       "      <td>{2: 47}</td>\n",
       "      <td>[8, 8]</td>\n",
       "      <td>{8: 59}</td>\n",
       "      <td>[6, 7, 8]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13182</th>\n",
       "      <td>chr21_37597069</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{4: 47}</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{16: 49}</td>\n",
       "      <td>[16, 19, 21]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14097</th>\n",
       "      <td>chr21_39337023</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>[15, 15]</td>\n",
       "      <td>{3: 39}</td>\n",
       "      <td>[15, 15]</td>\n",
       "      <td>{15: 39}</td>\n",
       "      <td>[15, 16, 22]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14530</th>\n",
       "      <td>chr21_40106894</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>[10, 10]</td>\n",
       "      <td>{2: 44}</td>\n",
       "      <td>[10, 10]</td>\n",
       "      <td>{10: 50}</td>\n",
       "      <td>[9, 10, 12]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15425</th>\n",
       "      <td>chr21_41776804</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{4: 45}</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>{16: 47}</td>\n",
       "      <td>[16, 17, 18]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               str_id  period  ref genotype_gangstr frequencies_gangstr  \\\n",
       "937    chr21_15994635       4    5         [20, 20]             {5: 45}   \n",
       "2768   chr21_19488881       2   22         [44, 44]            {22: 32}   \n",
       "3540   chr21_20982238       2   11         [22, 22]            {11: 32}   \n",
       "5994   chr21_25572109       2   11         [22, 22]            {11: 36}   \n",
       "7300   chr21_28173037       2   12         [24, 24]     {12: 27, 13: 7}   \n",
       "11671  chr21_35446085       2   15         [30, 30]            {15: 26}   \n",
       "12415  chr21_36462935       4    2           [8, 8]             {2: 47}   \n",
       "13182  chr21_37597069       4    4         [16, 16]             {4: 47}   \n",
       "14097  chr21_39337023       5    3         [15, 15]             {3: 39}   \n",
       "14530  chr21_40106894       5    2         [10, 10]             {2: 44}   \n",
       "15425  chr21_41776804       4    4         [16, 16]             {4: 45}   \n",
       "\n",
       "      genotype_hipstr frequencies_hipstr  genotype_sim  n_distinct_alleles  \\\n",
       "937          [20, 20]           {20: 47}  [20, 33, 37]                   3   \n",
       "2768         [44, 44]           {44: 35}  [35, 37, 44]                   3   \n",
       "3540         [22, 22]           {22: 45}  [13, 22, 25]                   3   \n",
       "5994         [22, 22]           {22: 50}  [22, 23, 25]                   3   \n",
       "7300         [24, 24]   {24: 42, 26: 10}  [24, 26, 27]                   3   \n",
       "11671        [30, 30]           {30: 40}  [27, 29, 30]                   3   \n",
       "12415          [8, 8]            {8: 59}     [6, 7, 8]                   3   \n",
       "13182        [16, 16]           {16: 49}  [16, 19, 21]                   3   \n",
       "14097        [15, 15]           {15: 39}  [15, 16, 22]                   3   \n",
       "14530        [10, 10]           {10: 50}   [9, 10, 12]                   3   \n",
       "15425        [16, 16]           {16: 47}  [16, 17, 18]                   3   \n",
       "\n",
       "       distinct_hipstr  \n",
       "937                  1  \n",
       "2768                 1  \n",
       "3540                 1  \n",
       "5994                 1  \n",
       "7300                 1  \n",
       "11671                1  \n",
       "12415                1  \n",
       "13182                1  \n",
       "14097                1  \n",
       "14530                1  \n",
       "15425                1  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_comparison.dropna(subset=\"frequencies_hipstr\").query(\"n_distinct_alleles == 3\").assign(\n",
    "    distinct_hipstr = lambda x: [len(set(i)) for i in x[\"genotype_hipstr\"]]\n",
    ").query(\"distinct_hipstr == 1\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
