{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12cd304d-31c1-492e-9f2a-692e9c6d7274",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "# SRC_DIR environment variable should be the absolute path to the 'ConSTRain-analyses' directory\n",
    "sys.path.append(os.environ[\"SRC_DIR\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ae5384da-70e5-42f1-81ef-f08ffcd3cf2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from multicopy_STR_genotyping import file_io\n",
    "\n",
    "sns.set_context(\"poster\")\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d66b0601-76e6-4d2b-b1f1-d0aea2f1ea6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf = \"../../../data/HG002/variants/ConSTRain/v0.9.1/HG002.GRCh38.2x250_depth10x_with_mononuc.vcf\" # str: path to input vcf file                      \n",
    "vcf_format = \"ConSTRain\" # str: which tool was used to genotype\n",
    "sample = \"HG002.GRCh38.2x250_depth10x\" # str: name of sample in vcf to analyse\n",
    "set_cn = False # bool: set CN of loci manually (based on chromosome name)\n",
    "gt_in_bp = False # bool: does the tool report genotypes in bp (True) or in number of repeat units (False)\n",
    "strs_covered_by_haplotypes = \"../../../data/HG002/regions/hg38_ver13_0boe_mononucleotides_union.bed\" # str: path to bed with loci that are covered by the haplotypes provided by Q100\n",
    "haplotype_str_length = \"../../../data/HG002/variants/HG002_GRCh38_STR_lengths.csv\" # str: path to csv with STR allele lengths based on Q100 haplotypes\n",
    "comparison_file_out = None # None or str: path to write csv with comparison of STR allele lengths from vcf to Q100 STR allele lengths\n",
    "\n",
    "# depth_filter = 0.05 # None or float. Alpha to use for filtering loci with extreme depth values\n",
    "lower, upper = 1.5, 6.5\n",
    "filt_segdup = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d168c683-ab62-42a2-aff0-c6c989fb02d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_cn_manually(df):\n",
    "    autosomes = [f\"chr{i}\" for i in range(1, 23)]\n",
    "    df = df.assign(\n",
    "        copy_number = lambda x: [2 if i in autosomes else 1 for i in x[\"chr\"]]\n",
    "    )\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3a137dd7-771e-4677-bd06-41589033c202",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1733646, 7) (1453440, 8)\n"
     ]
    }
   ],
   "source": [
    "df_repeats, df_hg002 = file_io.dfs_from_vcf(filename=vcf, samples=[sample], vcf_format=vcf_format)\n",
    "df_hg002 = df_hg002.dropna().assign(\n",
    "    depth = lambda x: [sum(i.values()) for i in x[\"frequencies\"]],\n",
    "    depth_norm = lambda x: x[\"depth\"] / x[\"copy_number\"]\n",
    ").reset_index(drop=True)\n",
    "df_hg002 = df_hg002.merge(df_repeats[[\"str_id\", \"period\"]], on=\"str_id\", how=\"left\")\n",
    "print(df_repeats.shape, df_hg002.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5eca5351-702d-4df4-8b1b-adfec85f1aef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1453440, 8)\n",
      "(1266055, 8)\n"
     ]
    }
   ],
   "source": [
    "if lower and upper:\n",
    "    print(df_hg002.shape)\n",
    "    df_hg002 = df_hg002.query(f\"depth_norm >= {lower} and depth_norm <= {upper}\").reset_index(drop=True)\n",
    "    print(df_hg002.shape)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "addac662-8c60-49c2-b283-26eb31e53510",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1695865, 7) (1249527, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>str_id</th>\n",
       "      <th>copy_number</th>\n",
       "      <th>frequencies</th>\n",
       "      <th>genotype</th>\n",
       "      <th>depth</th>\n",
       "      <th>depth_norm</th>\n",
       "      <th>period</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>unit</th>\n",
       "      <th>ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chr1_590659</td>\n",
       "      <td>2</td>\n",
       "      <td>{3: 5}</td>\n",
       "      <td>[3, 3]</td>\n",
       "      <td>5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>590659</td>\n",
       "      <td>590670</td>\n",
       "      <td>AAAT</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chr1_590969</td>\n",
       "      <td>2</td>\n",
       "      <td>{4: 5}</td>\n",
       "      <td>[4, 4]</td>\n",
       "      <td>5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>590969</td>\n",
       "      <td>590984</td>\n",
       "      <td>AAAC</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chr1_594083</td>\n",
       "      <td>2</td>\n",
       "      <td>{4: 3}</td>\n",
       "      <td>[4, 4]</td>\n",
       "      <td>3</td>\n",
       "      <td>1.5</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>594083</td>\n",
       "      <td>594094</td>\n",
       "      <td>TCC</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chr1_597686</td>\n",
       "      <td>2</td>\n",
       "      <td>{14: 4}</td>\n",
       "      <td>[14, 14]</td>\n",
       "      <td>4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>597686</td>\n",
       "      <td>597699</td>\n",
       "      <td>A</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chr1_599683</td>\n",
       "      <td>2</td>\n",
       "      <td>{5: 10}</td>\n",
       "      <td>[5, 5]</td>\n",
       "      <td>10</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>599683</td>\n",
       "      <td>599707</td>\n",
       "      <td>TTTTG</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1249522</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chrY_26619402</td>\n",
       "      <td>1</td>\n",
       "      <td>{12: 2}</td>\n",
       "      <td>[12]</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26619402</td>\n",
       "      <td>26619413</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1249523</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chrY_26619856</td>\n",
       "      <td>1</td>\n",
       "      <td>{10: 2}</td>\n",
       "      <td>[10]</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26619856</td>\n",
       "      <td>26619865</td>\n",
       "      <td>T</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1249524</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chrY_26622472</td>\n",
       "      <td>1</td>\n",
       "      <td>{2: 4}</td>\n",
       "      <td>[2]</td>\n",
       "      <td>4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26622472</td>\n",
       "      <td>26622481</td>\n",
       "      <td>AAATA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1249525</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chrY_26625608</td>\n",
       "      <td>1</td>\n",
       "      <td>{2: 2}</td>\n",
       "      <td>[2]</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26625608</td>\n",
       "      <td>26625621</td>\n",
       "      <td>TAAGTAT</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1249526</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chrY_26637436</td>\n",
       "      <td>1</td>\n",
       "      <td>{4: 5}</td>\n",
       "      <td>[4]</td>\n",
       "      <td>5</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26637436</td>\n",
       "      <td>26637447</td>\n",
       "      <td>TTA</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1249527 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              sample         str_id  copy_number frequencies  \\\n",
       "0        HG002.GRCh38.2x250_depth10x    chr1_590659            2      {3: 5}   \n",
       "1        HG002.GRCh38.2x250_depth10x    chr1_590969            2      {4: 5}   \n",
       "2        HG002.GRCh38.2x250_depth10x    chr1_594083            2      {4: 3}   \n",
       "3        HG002.GRCh38.2x250_depth10x    chr1_597686            2     {14: 4}   \n",
       "4        HG002.GRCh38.2x250_depth10x    chr1_599683            2     {5: 10}   \n",
       "...                              ...            ...          ...         ...   \n",
       "1249522  HG002.GRCh38.2x250_depth10x  chrY_26619402            1     {12: 2}   \n",
       "1249523  HG002.GRCh38.2x250_depth10x  chrY_26619856            1     {10: 2}   \n",
       "1249524  HG002.GRCh38.2x250_depth10x  chrY_26622472            1      {2: 4}   \n",
       "1249525  HG002.GRCh38.2x250_depth10x  chrY_26625608            1      {2: 2}   \n",
       "1249526  HG002.GRCh38.2x250_depth10x  chrY_26637436            1      {4: 5}   \n",
       "\n",
       "         genotype  depth  depth_norm  period   chr     start       end  \\\n",
       "0          [3, 3]      5         2.5       4  chr1    590659    590670   \n",
       "1          [4, 4]      5         2.5       4  chr1    590969    590984   \n",
       "2          [4, 4]      3         1.5       3  chr1    594083    594094   \n",
       "3        [14, 14]      4         2.0       1  chr1    597686    597699   \n",
       "4          [5, 5]     10         5.0       5  chr1    599683    599707   \n",
       "...           ...    ...         ...     ...   ...       ...       ...   \n",
       "1249522      [12]      2         2.0       1  chrY  26619402  26619413   \n",
       "1249523      [10]      2         2.0       1  chrY  26619856  26619865   \n",
       "1249524       [2]      4         4.0       5  chrY  26622472  26622481   \n",
       "1249525       [2]      2         2.0       7  chrY  26625608  26625621   \n",
       "1249526       [4]      5         5.0       3  chrY  26637436  26637447   \n",
       "\n",
       "            unit  ref  \n",
       "0           AAAT    3  \n",
       "1           AAAC    4  \n",
       "2            TCC    4  \n",
       "3              A   14  \n",
       "4          TTTTG    5  \n",
       "...          ...  ...  \n",
       "1249522        A   12  \n",
       "1249523        T   10  \n",
       "1249524    AAATA    2  \n",
       "1249525  TAAGTAT    2  \n",
       "1249526      TTA    4  \n",
       "\n",
       "[1249527 rows x 13 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_covered_repeats = pd.read_csv(strs_covered_by_haplotypes, sep=\"\\t\", names=[\"chr\", \"start\", \"end\", \"period\", \"unit\"])\n",
    "df_covered_repeats = df_covered_repeats.assign(\n",
    "            str_id = np.array([f\"{chrom}_{start + 1}\" for chrom, start in zip(df_covered_repeats.chr, df_covered_repeats.start)])\n",
    ")\n",
    "\n",
    "df_covered_repeats = df_repeats[df_repeats[\"str_id\"].isin(df_covered_repeats[\"str_id\"])]\n",
    "\n",
    "df_hg002 = df_hg002.merge(\n",
    "    df_covered_repeats[[\"str_id\", \"chr\", \"start\", \"end\", \"unit\", \"ref\"]], on=\"str_id\", how=\"inner\"\n",
    ")\n",
    "\n",
    "if set_cn:\n",
    "    df_hg002 = set_cn_manually(df_hg002)\n",
    "    \n",
    "df_hg002 = df_hg002.dropna().reset_index(drop=True)\n",
    "\n",
    "print(df_covered_repeats.shape, df_hg002.shape)\n",
    "\n",
    "df_hg002\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2e5f441d-5e41-4613-9c7f-41ec15728761",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1202944, 13)\n"
     ]
    }
   ],
   "source": [
    "if filt_segdup:\n",
    "    df_in_segdup = pd.read_csv(\"../../../data/hg38_ver13_0boe_mononucleotides_in_segdup.bed\", sep=\"\\t\", names=[\"chr\", \"start\", \"end\", \"period\", \"unit\"])\n",
    "    df_in_segdup = df_in_segdup.assign(str_id = lambda x: [f\"{i}_{j + 1}\" for i, j in zip(x[\"chr\"], x[\"start\"])])\n",
    "    \n",
    "    df_hg002 = df_hg002[~df_hg002[\"str_id\"].isin(df_in_segdup[\"str_id\"])].reset_index(drop=True)\n",
    "    print(df_hg002.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a22aa01e-e353-4c58-8251-86c75025f7d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1202944, 13)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "70.93394816214735"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(df_hg002.shape)\n",
    "df_hg002.shape[0] / df_covered_repeats.shape[0] * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "32f504da-3262-4d81-ba22-8d267b921546",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>str_id</th>\n",
       "      <th>copy_number</th>\n",
       "      <th>frequencies</th>\n",
       "      <th>genotype</th>\n",
       "      <th>depth</th>\n",
       "      <th>depth_norm</th>\n",
       "      <th>period</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>unit</th>\n",
       "      <th>ref</th>\n",
       "      <th>illumina_region_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chr1_904440</td>\n",
       "      <td>2</td>\n",
       "      <td>{3: 8}</td>\n",
       "      <td>[3, 3]</td>\n",
       "      <td>8</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>904440</td>\n",
       "      <td>904448</td>\n",
       "      <td>CTC</td>\n",
       "      <td>3</td>\n",
       "      <td>[9, 9]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chr1_904654</td>\n",
       "      <td>2</td>\n",
       "      <td>{4: 7}</td>\n",
       "      <td>[4, 4]</td>\n",
       "      <td>7</td>\n",
       "      <td>3.5</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>904654</td>\n",
       "      <td>904665</td>\n",
       "      <td>CCT</td>\n",
       "      <td>4</td>\n",
       "      <td>[12, 12]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chr1_914949</td>\n",
       "      <td>2</td>\n",
       "      <td>{16: 8}</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>8</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>914949</td>\n",
       "      <td>914964</td>\n",
       "      <td>A</td>\n",
       "      <td>16</td>\n",
       "      <td>[16, 16]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chr1_919012</td>\n",
       "      <td>2</td>\n",
       "      <td>{3: 8}</td>\n",
       "      <td>[3, 3]</td>\n",
       "      <td>8</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>919012</td>\n",
       "      <td>919023</td>\n",
       "      <td>GCAG</td>\n",
       "      <td>3</td>\n",
       "      <td>[12, 12]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chr1_919291</td>\n",
       "      <td>2</td>\n",
       "      <td>{2: 8}</td>\n",
       "      <td>[2, 2]</td>\n",
       "      <td>8</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>919291</td>\n",
       "      <td>919300</td>\n",
       "      <td>CACCC</td>\n",
       "      <td>2</td>\n",
       "      <td>[10, 10]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202939</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chrY_26474096</td>\n",
       "      <td>1</td>\n",
       "      <td>{3: 3}</td>\n",
       "      <td>[3]</td>\n",
       "      <td>3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26474096</td>\n",
       "      <td>26474107</td>\n",
       "      <td>CAAA</td>\n",
       "      <td>3</td>\n",
       "      <td>[12]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202940</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chrY_26476426</td>\n",
       "      <td>1</td>\n",
       "      <td>{14: 2}</td>\n",
       "      <td>[14]</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26476426</td>\n",
       "      <td>26476439</td>\n",
       "      <td>T</td>\n",
       "      <td>14</td>\n",
       "      <td>[14]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202941</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chrY_26487819</td>\n",
       "      <td>1</td>\n",
       "      <td>{10: 2}</td>\n",
       "      <td>[10]</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26487819</td>\n",
       "      <td>26487828</td>\n",
       "      <td>A</td>\n",
       "      <td>10</td>\n",
       "      <td>[10]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202942</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chrY_26487987</td>\n",
       "      <td>1</td>\n",
       "      <td>{10: 5}</td>\n",
       "      <td>[10]</td>\n",
       "      <td>5</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26487987</td>\n",
       "      <td>26487996</td>\n",
       "      <td>A</td>\n",
       "      <td>10</td>\n",
       "      <td>[10]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202943</th>\n",
       "      <td>HG002.GRCh38.2x250_depth10x</td>\n",
       "      <td>chrY_26489935</td>\n",
       "      <td>1</td>\n",
       "      <td>{4: 4}</td>\n",
       "      <td>[4]</td>\n",
       "      <td>4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26489935</td>\n",
       "      <td>26489946</td>\n",
       "      <td>AGA</td>\n",
       "      <td>4</td>\n",
       "      <td>[12]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1202944 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              sample         str_id  copy_number frequencies  \\\n",
       "0        HG002.GRCh38.2x250_depth10x    chr1_904440            2      {3: 8}   \n",
       "1        HG002.GRCh38.2x250_depth10x    chr1_904654            2      {4: 7}   \n",
       "2        HG002.GRCh38.2x250_depth10x    chr1_914949            2     {16: 8}   \n",
       "3        HG002.GRCh38.2x250_depth10x    chr1_919012            2      {3: 8}   \n",
       "4        HG002.GRCh38.2x250_depth10x    chr1_919291            2      {2: 8}   \n",
       "...                              ...            ...          ...         ...   \n",
       "1202939  HG002.GRCh38.2x250_depth10x  chrY_26474096            1      {3: 3}   \n",
       "1202940  HG002.GRCh38.2x250_depth10x  chrY_26476426            1     {14: 2}   \n",
       "1202941  HG002.GRCh38.2x250_depth10x  chrY_26487819            1     {10: 2}   \n",
       "1202942  HG002.GRCh38.2x250_depth10x  chrY_26487987            1     {10: 5}   \n",
       "1202943  HG002.GRCh38.2x250_depth10x  chrY_26489935            1      {4: 4}   \n",
       "\n",
       "         genotype  depth  depth_norm  period   chr     start       end   unit  \\\n",
       "0          [3, 3]      8         4.0       3  chr1    904440    904448    CTC   \n",
       "1          [4, 4]      7         3.5       3  chr1    904654    904665    CCT   \n",
       "2        [16, 16]      8         4.0       1  chr1    914949    914964      A   \n",
       "3          [3, 3]      8         4.0       4  chr1    919012    919023   GCAG   \n",
       "4          [2, 2]      8         4.0       5  chr1    919291    919300  CACCC   \n",
       "...           ...    ...         ...     ...   ...       ...       ...    ...   \n",
       "1202939       [3]      3         3.0       4  chrY  26474096  26474107   CAAA   \n",
       "1202940      [14]      2         2.0       1  chrY  26476426  26476439      T   \n",
       "1202941      [10]      2         2.0       1  chrY  26487819  26487828      A   \n",
       "1202942      [10]      5         5.0       1  chrY  26487987  26487996      A   \n",
       "1202943       [4]      4         4.0       3  chrY  26489935  26489946    AGA   \n",
       "\n",
       "         ref illumina_region_len  \n",
       "0          3              [9, 9]  \n",
       "1          4            [12, 12]  \n",
       "2         16            [16, 16]  \n",
       "3          3            [12, 12]  \n",
       "4          2            [10, 10]  \n",
       "...      ...                 ...  \n",
       "1202939    3                [12]  \n",
       "1202940   14                [14]  \n",
       "1202941   10                [10]  \n",
       "1202942   10                [10]  \n",
       "1202943    4                [12]  \n",
       "\n",
       "[1202944 rows x 14 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if gt_in_bp:\n",
    "    df_hg002 = df_hg002.assign(\n",
    "        illumina_region_len = lambda x: [sorted(gt) for gt in x[\"genotype\"]],\n",
    "        depth = lambda x: [sum(i.values()) for i in x[\"frequencies\"]],\n",
    "    )\n",
    "else:\n",
    "    df_hg002 = df_hg002.assign(\n",
    "        illumina_region_len = lambda x: [sorted([k * j for k in i]) for i, j in zip(x[\"genotype\"], x[\"period\"])],\n",
    "        depth = lambda x: [sum(i.values()) for i in x[\"frequencies\"]],\n",
    "    )\n",
    "    \n",
    "df_hg002"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b40f6e52-73ae-4131-8d27-da8502f0c066",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_id</th>\n",
       "      <th>region_len_ref</th>\n",
       "      <th>region_len_mat</th>\n",
       "      <th>region_len_pat</th>\n",
       "      <th>chr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_588068</td>\n",
       "      <td>12</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_589245</td>\n",
       "      <td>12</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_590659</td>\n",
       "      <td>12</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_590969</td>\n",
       "      <td>16</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_591734</td>\n",
       "      <td>18</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695860</th>\n",
       "      <td>chrY_57188868</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "      <td>chrY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695861</th>\n",
       "      <td>chrY_57200838</td>\n",
       "      <td>14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.0</td>\n",
       "      <td>chrY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695862</th>\n",
       "      <td>chrY_57201000</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.0</td>\n",
       "      <td>chrY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695863</th>\n",
       "      <td>chrY_57202381</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23.0</td>\n",
       "      <td>chrY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695864</th>\n",
       "      <td>chrY_57202848</td>\n",
       "      <td>26</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28.0</td>\n",
       "      <td>chrY</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1695865 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                str_id  region_len_ref  region_len_mat  region_len_pat   chr\n",
       "0          chr1_588068              12            12.0            12.0  chr1\n",
       "1          chr1_589245              12            12.0            12.0  chr1\n",
       "2          chr1_590659              12            12.0            12.0  chr1\n",
       "3          chr1_590969              16            16.0            16.0  chr1\n",
       "4          chr1_591734              18            20.0            20.0  chr1\n",
       "...                ...             ...             ...             ...   ...\n",
       "1695860  chrY_57188868              11             NaN            11.0  chrY\n",
       "1695861  chrY_57200838              14             NaN            15.0  chrY\n",
       "1695862  chrY_57201000              20             NaN            20.0  chrY\n",
       "1695863  chrY_57202381              23             NaN            23.0  chrY\n",
       "1695864  chrY_57202848              26             NaN            28.0  chrY\n",
       "\n",
       "[1695865 rows x 5 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_strs_haplotypes = pd.read_csv(haplotype_str_length).assign(chr = lambda x: [i[0] for i in x[\"str_id\"].str.split(\"_\")])\n",
    "df_strs_haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b7fb63b2-8c61-45e6-a6c9-9ec20c9436a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_id</th>\n",
       "      <th>region_len_ref</th>\n",
       "      <th>region_len_mat</th>\n",
       "      <th>region_len_pat</th>\n",
       "      <th>chr</th>\n",
       "      <th>haplo_region_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_588068</td>\n",
       "      <td>12</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[12, 12]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_589245</td>\n",
       "      <td>12</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[12, 12]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_590659</td>\n",
       "      <td>12</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[12, 12]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_590969</td>\n",
       "      <td>16</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[16, 16]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_591734</td>\n",
       "      <td>18</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>[20, 20]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695860</th>\n",
       "      <td>chrY_57188868</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "      <td>chrY</td>\n",
       "      <td>[11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695861</th>\n",
       "      <td>chrY_57200838</td>\n",
       "      <td>14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.0</td>\n",
       "      <td>chrY</td>\n",
       "      <td>[15]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695862</th>\n",
       "      <td>chrY_57201000</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.0</td>\n",
       "      <td>chrY</td>\n",
       "      <td>[20]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695863</th>\n",
       "      <td>chrY_57202381</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23.0</td>\n",
       "      <td>chrY</td>\n",
       "      <td>[23]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695864</th>\n",
       "      <td>chrY_57202848</td>\n",
       "      <td>26</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28.0</td>\n",
       "      <td>chrY</td>\n",
       "      <td>[28]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1695865 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                str_id  region_len_ref  region_len_mat  region_len_pat   chr  \\\n",
       "0          chr1_588068              12            12.0            12.0  chr1   \n",
       "1          chr1_589245              12            12.0            12.0  chr1   \n",
       "2          chr1_590659              12            12.0            12.0  chr1   \n",
       "3          chr1_590969              16            16.0            16.0  chr1   \n",
       "4          chr1_591734              18            20.0            20.0  chr1   \n",
       "...                ...             ...             ...             ...   ...   \n",
       "1695860  chrY_57188868              11             NaN            11.0  chrY   \n",
       "1695861  chrY_57200838              14             NaN            15.0  chrY   \n",
       "1695862  chrY_57201000              20             NaN            20.0  chrY   \n",
       "1695863  chrY_57202381              23             NaN            23.0  chrY   \n",
       "1695864  chrY_57202848              26             NaN            28.0  chrY   \n",
       "\n",
       "        haplo_region_len  \n",
       "0               [12, 12]  \n",
       "1               [12, 12]  \n",
       "2               [12, 12]  \n",
       "3               [16, 16]  \n",
       "4               [20, 20]  \n",
       "...                  ...  \n",
       "1695860             [11]  \n",
       "1695861             [15]  \n",
       "1695862             [20]  \n",
       "1695863             [23]  \n",
       "1695864             [28]  \n",
       "\n",
       "[1695865 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "autosomes = \"chr != 'chrX' and chr != 'chrY'\"\n",
    "X = \"chr == 'chrX'\"\n",
    "Y = \"chr == 'chrY'\"\n",
    "\n",
    "df_strs_haplotypes[\"haplo_region_len\"] = [[] for i in range(len(df_strs_haplotypes))]\n",
    "\n",
    "df_strs_haplotypes.loc[df_strs_haplotypes.eval(autosomes), \"haplo_region_len\"] = pd.Series([\n",
    "    sorted([np.int64(i), np.int64(j)]) for i, j in zip(\n",
    "        df_strs_haplotypes.loc[df_strs_haplotypes.eval(autosomes), \"region_len_mat\"], \n",
    "        df_strs_haplotypes.loc[df_strs_haplotypes.eval(autosomes), \"region_len_pat\"],)])\n",
    "\n",
    "for i, j in zip(\n",
    "        df_strs_haplotypes.loc[df_strs_haplotypes.eval(X), \"haplo_region_len\"],\n",
    "        df_strs_haplotypes.loc[df_strs_haplotypes.eval(X), \"region_len_mat\"]):\n",
    "    i.append(np.int64(j))\n",
    "    \n",
    "for i, j in zip(\n",
    "        df_strs_haplotypes.loc[df_strs_haplotypes.eval(Y), \"haplo_region_len\"],\n",
    "        df_strs_haplotypes.loc[df_strs_haplotypes.eval(Y), \"region_len_pat\"]):\n",
    "    i.append(np.int64(j))\n",
    "    \n",
    "df_strs_haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5802936d-93d0-4b9b-b345-0b9d12d845a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_id</th>\n",
       "      <th>haplo_region_len</th>\n",
       "      <th>illumina_region_len</th>\n",
       "      <th>depth</th>\n",
       "      <th>copy_number</th>\n",
       "      <th>period</th>\n",
       "      <th>match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_904440</td>\n",
       "      <td>[9, 9]</td>\n",
       "      <td>[9, 9]</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_904654</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_914949</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_919012</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>[12, 12]</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_919291</td>\n",
       "      <td>[10, 10]</td>\n",
       "      <td>[10, 10]</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202939</th>\n",
       "      <td>chrY_26474096</td>\n",
       "      <td>[12]</td>\n",
       "      <td>[12]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202940</th>\n",
       "      <td>chrY_26476426</td>\n",
       "      <td>[14]</td>\n",
       "      <td>[14]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202941</th>\n",
       "      <td>chrY_26487819</td>\n",
       "      <td>[10]</td>\n",
       "      <td>[10]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202942</th>\n",
       "      <td>chrY_26487987</td>\n",
       "      <td>[10]</td>\n",
       "      <td>[10]</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202943</th>\n",
       "      <td>chrY_26489935</td>\n",
       "      <td>[12]</td>\n",
       "      <td>[12]</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1202944 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                str_id haplo_region_len illumina_region_len  depth  \\\n",
       "0          chr1_904440           [9, 9]              [9, 9]      8   \n",
       "1          chr1_904654         [12, 12]            [12, 12]      7   \n",
       "2          chr1_914949         [16, 16]            [16, 16]      8   \n",
       "3          chr1_919012         [12, 12]            [12, 12]      8   \n",
       "4          chr1_919291         [10, 10]            [10, 10]      8   \n",
       "...                ...              ...                 ...    ...   \n",
       "1202939  chrY_26474096             [12]                [12]      3   \n",
       "1202940  chrY_26476426             [14]                [14]      2   \n",
       "1202941  chrY_26487819             [10]                [10]      2   \n",
       "1202942  chrY_26487987             [10]                [10]      5   \n",
       "1202943  chrY_26489935             [12]                [12]      4   \n",
       "\n",
       "         copy_number  period  match  \n",
       "0                  2       3   True  \n",
       "1                  2       3   True  \n",
       "2                  2       1   True  \n",
       "3                  2       4   True  \n",
       "4                  2       5   True  \n",
       "...              ...     ...    ...  \n",
       "1202939            1       4   True  \n",
       "1202940            1       1   True  \n",
       "1202941            1       1   True  \n",
       "1202942            1       1   True  \n",
       "1202943            1       3   True  \n",
       "\n",
       "[1202944 rows x 7 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_comparison = (\n",
    "    df_strs_haplotypes[[\"str_id\", \"haplo_region_len\"]]\n",
    "        .merge(df_hg002[[\"str_id\", \"illumina_region_len\", \"depth\", \"copy_number\", \"period\"]], on=\"str_id\", how=\"inner\")\n",
    "        .assign(\n",
    "            match = lambda x: x[\"haplo_region_len\"] == x[\"illumina_region_len\"],\n",
    "        )\n",
    ")\n",
    "\n",
    "df_comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b5051a0a-ad35-4664-b1a0-03c207bb7f14",
   "metadata": {},
   "outputs": [],
   "source": [
    "if comparison_file_out:\n",
    "    df_comparison.to_csv(comparison_file_out, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ac898654-98c5-4fed-9c19-4b67bbae5e27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1431554"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1332322 + 99232"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c2e5f48d-2dd6-4ed3-9631-ff671ce033f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n correct: 1139731, n wrong: 63213, n total: 1202944\n",
      "Overall accuracy: 94.74514191849329\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_id</th>\n",
       "      <th>haplo_region_len</th>\n",
       "      <th>illumina_region_len</th>\n",
       "      <th>depth</th>\n",
       "      <th>copy_number</th>\n",
       "      <th>period</th>\n",
       "      <th>match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>chr1_1020848</td>\n",
       "      <td>[13, 14]</td>\n",
       "      <td>[13, 13]</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>chr1_1081643</td>\n",
       "      <td>[11, 12]</td>\n",
       "      <td>[10, 10]</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>chr1_1293732</td>\n",
       "      <td>[60, 80]</td>\n",
       "      <td>[60, 60]</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>chr1_1363439</td>\n",
       "      <td>[13, 13]</td>\n",
       "      <td>[15, 15]</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>chr1_1369036</td>\n",
       "      <td>[16, 16]</td>\n",
       "      <td>[16, 17]</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1200369</th>\n",
       "      <td>chrY_7263346</td>\n",
       "      <td>[15]</td>\n",
       "      <td>[14]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201340</th>\n",
       "      <td>chrY_14296235</td>\n",
       "      <td>[38]</td>\n",
       "      <td>[36]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201368</th>\n",
       "      <td>chrY_14409113</td>\n",
       "      <td>[15]</td>\n",
       "      <td>[16]</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201513</th>\n",
       "      <td>chrY_14892739</td>\n",
       "      <td>[13]</td>\n",
       "      <td>[14]</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202282</th>\n",
       "      <td>chrY_18944110</td>\n",
       "      <td>[10]</td>\n",
       "      <td>[11]</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63213 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                str_id haplo_region_len illumina_region_len  depth  \\\n",
       "54        chr1_1020848         [13, 14]            [13, 13]      3   \n",
       "72        chr1_1081643         [11, 12]            [10, 10]      3   \n",
       "153       chr1_1293732         [60, 80]            [60, 60]      5   \n",
       "189       chr1_1363439         [13, 13]            [15, 15]      7   \n",
       "192       chr1_1369036         [16, 16]            [16, 17]      7   \n",
       "...                ...              ...                 ...    ...   \n",
       "1200369   chrY_7263346             [15]                [14]      3   \n",
       "1201340  chrY_14296235             [38]                [36]      2   \n",
       "1201368  chrY_14409113             [15]                [16]      6   \n",
       "1201513  chrY_14892739             [13]                [14]      2   \n",
       "1202282  chrY_18944110             [10]                [11]      3   \n",
       "\n",
       "         copy_number  period  match  \n",
       "54                 2       1  False  \n",
       "72                 2       5  False  \n",
       "153                2      20  False  \n",
       "189                2       1  False  \n",
       "192                2       1  False  \n",
       "...              ...     ...    ...  \n",
       "1200369            1       1  False  \n",
       "1201340            1       2  False  \n",
       "1201368            1       1  False  \n",
       "1201513            1       1  False  \n",
       "1202282            1       1  False  \n",
       "\n",
       "[63213 rows x 7 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_wrong = df_comparison.loc[~df_comparison[\"match\"]].shape[0]\n",
    "n_total = df_comparison.shape[0]\n",
    "\n",
    "print(f\"n correct: {n_total - n_wrong}, n wrong: {n_wrong}, n total: {n_total}\")\n",
    "print(\"Overall accuracy:\", (1 - (n_wrong / n_total)) * 100)\n",
    "\n",
    "df_comparison.loc[~df_comparison[\"match\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3dfee9f-e4aa-4948-8235-f762e044119e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_plot = pd.DataFrame(\n",
    "    np.concatenate([np.repeat([1, 2, 3, 4, 5, 6], 7).reshape(-1, 1), np.tile([1, 5, 5, 10, 10, 15, 15, 20, 20, 25, 25, 30, 30, 1e6], 6).reshape(-1, 2)], axis=1),\n",
    "    columns=[\"period\", \"depth_lower\", \"depth_upper\"]\n",
    ")\n",
    "\n",
    "accuracy = []\n",
    "n = []\n",
    "for row in df_plot.to_dict(orient=\"records\"):\n",
    "    selected_loci = (\n",
    "        df_comparison\n",
    "            .assign(depth = lambda x: x[\"depth\"] / x[\"copy_number\"])\n",
    "            .query(f\"depth >= {row['depth_lower']} and depth < {row['depth_upper']} and period == {row['period']}\")[\"str_id\"]\n",
    "    )\n",
    "    df_comparison_filt = df_comparison.loc[df_comparison[\"str_id\"].isin(selected_loci)]\n",
    "    current_n = df_comparison_filt.shape[0]\n",
    "    if current_n == 0:\n",
    "        n.append(np.nan)\n",
    "        accuracy.append(np.nan)\n",
    "    else:\n",
    "        n.append(current_n)\n",
    "        accuracy.append((1 - (df_comparison_filt.loc[~df_comparison_filt[\"match\"]].shape[0] / df_comparison_filt.shape[0])) * 100)\n",
    "\n",
    "df_plot[\"n\"] = n\n",
    "df_plot[\"accuracy\"] = accuracy\n",
    "df_plot[\"depth\"] = np.tile([\"1-5\", \"5-10\", \"10-15\", \"15-20\", \"20-25\", \"25-30\", \"30-\"], 6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "491531cf-8ce7-4254-8dc7-a8ab7375974c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for name, data in df_plot.groupby(\"depth\"):\n",
    "    print(name, np.average(data[\"accuracy\"], weights=data[\"n\"]), data[\"n\"].sum())\n",
    "print()\n",
    "for name, data in df_plot.groupby(\"period\"):\n",
    "    print(name, np.average(data[\"accuracy\"], weights=data[\"n\"]), data[\"n\"].sum())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ee6690a-a862-40ae-ba40-6fe46d269838",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(10, 8))\n",
    "\n",
    "ax = sns.barplot(\n",
    "    df_plot,\n",
    "    x=\"period\",\n",
    "    y=\"accuracy\",\n",
    "    hue=\"depth\",\n",
    "    palette=\"colorblind\"\n",
    ")\n",
    "\n",
    "ax.set(\n",
    "    xlabel=\"Period\",\n",
    "    ylabel = \"Accuracy\"\n",
    ")\n",
    "\n",
    "_ = ax.get_legend().set(\n",
    "    title=\"Depth / CN\",\n",
    "    bbox_to_anchor = (1, 1)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98a278b4-e241-4662-9c8b-25b6e3bd1113",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import ndimage\n",
    "\n",
    "df_plot_density = (\n",
    "    df_comparison\n",
    "        .query(\"period >= 1 and period <= 6\")\n",
    "        .assign(depth = lambda x: x[\"depth\"] / x[\"copy_number\"])[[\"match\", \"period\", \"depth\"]]\n",
    ")\n",
    "\n",
    "df_plot_smooth = (\n",
    "    df_plot_density        \n",
    "        .drop(\"period\", axis=1)\n",
    "        .groupby([\"depth\"], observed=True, as_index=False).agg(\n",
    "            accuracy = (\"match\", lambda x: x.sum() / len(x)), \n",
    "            n_obs=(\"match\", \"count\"))\n",
    "        # .query(f\"n_obs >= {min_obs}\")\n",
    "        \n",
    ")\n",
    "\n",
    "smooth = []\n",
    "smooth = ndimage.gaussian_filter1d(df_plot_smooth[\"accuracy\"], 3, mode=\"reflect\")\n",
    "df_plot_smooth = df_plot_smooth.assign(smooth = smooth)\n",
    "\n",
    "\n",
    "df_plot_smooth_period = (\n",
    "    df_plot_density\n",
    "        .groupby([\"period\", \"depth\"], observed=True, as_index=False).agg(\n",
    "            accuracy = (\"match\", lambda x: x.sum() / len(x)), \n",
    "            n_obs=(\"match\", \"count\"))\n",
    "        # .query(f\"n_obs >= {min_obs}\")\n",
    "        \n",
    ")\n",
    "\n",
    "smooth = []\n",
    "for group, data in df_plot_smooth_period.groupby([\"period\"]):\n",
    "    current_smooth = ndimage.gaussian_filter1d(data[\"accuracy\"], 3, mode=\"reflect\")\n",
    "    smooth.append(current_smooth)\n",
    "df_plot_smooth_period = df_plot_smooth_period.assign(smooth = np.concatenate(smooth))\n",
    "\n",
    "df_plot_smooth_period"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f106cb3f-1b27-48ae-9c90-13b0ff4b020f",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(dpi=125)\n",
    "\n",
    "ax = sns.lineplot(\n",
    "    df_plot_smooth.assign(tmp=\"ConSTRain\"), # hack to add a legend\n",
    "    x = \"depth\",\n",
    "    y = \"smooth\",\n",
    "    hue=\"tmp\",\n",
    "    palette=\"colorblind\",\n",
    "    legend=False\n",
    ")\n",
    "\n",
    "ax = sns.scatterplot(\n",
    "    df_plot_smooth,\n",
    "    x = \"depth\",\n",
    "    y = \"accuracy\",\n",
    "    color=sns.color_palette(\"colorblind\")[0],\n",
    "    s=25,\n",
    "    ax = ax,\n",
    ")\n",
    "\n",
    "ax.set(\n",
    "    xlabel = \"Sequencing depth / CN\",\n",
    "    ylabel = \"Accuracy\",\n",
    "    ylim = (0, 1.05),\n",
    "    xlim = (0, 20)\n",
    ")\n",
    "ax.yaxis.label.set_color(color=sns.color_palette(\"colorblind\")[0])\n",
    "\n",
    "ax2 = ax.twinx()\n",
    "ax2 = sns.histplot(\n",
    "    df_plot_density,\n",
    "    x = \"depth\",\n",
    "    discrete=True,\n",
    "    stat=\"proportion\",\n",
    "    color=\"grey\",\n",
    "    ax=ax2,\n",
    ")\n",
    "ax2.set(\n",
    "    ylabel=\"Proportion of STR loci\",\n",
    "    # ylim=(0, 0.07),\n",
    ")\n",
    "ax2.yaxis.label.set_color(color=\"grey\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "472514ff-cce3-49f7-9adb-df79e915c4aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(3, 2, figsize=(20, 20))\n",
    "\n",
    "for period, ax in zip(range(1, 7), axes.ravel()):\n",
    "    df_subplot = (df_plot_smooth_period\n",
    "                  .query(f\"period == {period}\"))\n",
    "\n",
    "    ax = sns.lineplot(\n",
    "        df_subplot,\n",
    "        x = \"depth\",\n",
    "        y = \"smooth\",\n",
    "        color=sns.color_palette(\"colorblind\")[0],\n",
    "        linewidth=5,\n",
    "        ax = ax,\n",
    "    )\n",
    "\n",
    "    ax = sns.scatterplot(\n",
    "        df_subplot,\n",
    "        x = \"depth\",\n",
    "        y = \"accuracy\",\n",
    "        color=sns.color_palette(\"colorblind\")[0],\n",
    "        s=25,\n",
    "        ax = ax,\n",
    "    )\n",
    "\n",
    "    ax.set(\n",
    "        title = f\"Period {period}\",\n",
    "        xlabel = \"Sequencing depth / CN\",\n",
    "        ylabel = \"Accuracy\",\n",
    "        ylim = (0, 1.05),\n",
    "        xlim = (0, 60)\n",
    "    )\n",
    "    ax.yaxis.label.set_color(color=sns.color_palette(\"colorblind\")[0])\n",
    "    \n",
    "    ax2 = ax.twinx()\n",
    "    ax2 = sns.histplot(\n",
    "        df_plot_density.query(f\"period == {period}\"),\n",
    "        x = \"depth\",\n",
    "        discrete=True,\n",
    "        stat=\"proportion\",\n",
    "        color=\"grey\",\n",
    "        ax=ax2,\n",
    "    )\n",
    "    \n",
    "    ax2.set(\n",
    "        ylabel=\"Proportion of STR loci\",\n",
    "        ylim=(0, 0.08),\n",
    "    )\n",
    "    ax2.yaxis.label.set_color(color=\"grey\")\n",
    "\n",
    "plt.tight_layout()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
