{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d90bc349-9475-4139-8165-a85cfb04cad5",
   "metadata": {},
   "source": [
    "# Generating a new reference sequence that incorporates specific STR lengths\n",
    "\n",
    "We want to make a new reference sequence that is based on an existing reference, but has different STR allele lengths based on a list of STR loci provided."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c286372f-71d7-40df-a94a-4a7030819ab7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "# SRC_DIR environment variable should be the absolute path to the 'ConSTRain-analyses' directory\n",
    "sys.path.append(os.environ[\"SRC_DIR\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "047785c6-23ce-4fc9-91fb-a4ab23f7fca8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import pandas as pd\n",
    "import pysam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70d2dc34-7fc2-4e6d-aed1-8839455439bf",
   "metadata": {},
   "source": [
    "Load a panel with STR locus annotations for GRCh38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e8543e33-4ee3-45f5-9ecb-6edba5b5edcb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>period</th>\n",
       "      <th>unit</th>\n",
       "      <th>str_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>926634</th>\n",
       "      <td>chr21</td>\n",
       "      <td>5011211</td>\n",
       "      <td>5011223</td>\n",
       "      <td>4</td>\n",
       "      <td>AATC</td>\n",
       "      <td>chr21_5011212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926635</th>\n",
       "      <td>chr21</td>\n",
       "      <td>5013076</td>\n",
       "      <td>5013088</td>\n",
       "      <td>2</td>\n",
       "      <td>CA</td>\n",
       "      <td>chr21_5013077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926636</th>\n",
       "      <td>chr21</td>\n",
       "      <td>5014891</td>\n",
       "      <td>5014903</td>\n",
       "      <td>2</td>\n",
       "      <td>CA</td>\n",
       "      <td>chr21_5014892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926637</th>\n",
       "      <td>chr21</td>\n",
       "      <td>5016247</td>\n",
       "      <td>5016265</td>\n",
       "      <td>3</td>\n",
       "      <td>GAG</td>\n",
       "      <td>chr21_5016248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926638</th>\n",
       "      <td>chr21</td>\n",
       "      <td>5016788</td>\n",
       "      <td>5016798</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>chr21_5016789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>948133</th>\n",
       "      <td>chr21</td>\n",
       "      <td>46688831</td>\n",
       "      <td>46688863</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>chr21_46688832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>948134</th>\n",
       "      <td>chr21</td>\n",
       "      <td>46689711</td>\n",
       "      <td>46689723</td>\n",
       "      <td>4</td>\n",
       "      <td>ATTT</td>\n",
       "      <td>chr21_46689712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>948135</th>\n",
       "      <td>chr21</td>\n",
       "      <td>46696281</td>\n",
       "      <td>46696292</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>chr21_46696282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>948136</th>\n",
       "      <td>chr21</td>\n",
       "      <td>46697534</td>\n",
       "      <td>46697554</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>chr21_46697535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>948137</th>\n",
       "      <td>chr21</td>\n",
       "      <td>46699459</td>\n",
       "      <td>46699471</td>\n",
       "      <td>4</td>\n",
       "      <td>GGGC</td>\n",
       "      <td>chr21_46699460</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21504 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr     start       end  period  unit          str_id\n",
       "926634  chr21   5011211   5011223       4  AATC   chr21_5011212\n",
       "926635  chr21   5013076   5013088       2    CA   chr21_5013077\n",
       "926636  chr21   5014891   5014903       2    CA   chr21_5014892\n",
       "926637  chr21   5016247   5016265       3   GAG   chr21_5016248\n",
       "926638  chr21   5016788   5016798       1     G   chr21_5016789\n",
       "...       ...       ...       ...     ...   ...             ...\n",
       "948133  chr21  46688831  46688863       2    AC  chr21_46688832\n",
       "948134  chr21  46689711  46689723       4  ATTT  chr21_46689712\n",
       "948135  chr21  46696281  46696292       1     T  chr21_46696282\n",
       "948136  chr21  46697534  46697554       1     A  chr21_46697535\n",
       "948137  chr21  46699459  46699471       4  GGGC  chr21_46699460\n",
       "\n",
       "[21504 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "names = [\"chr\", \"start\", \"end\", \"period\", \"unit\"]\n",
    "df_repeats_grch38 = (pd.read_csv(\"../../data/hg38_ver13_0boe_mononucleotides.bed\", sep=\"\\t\", names=names)\n",
    "                         .query(\"chr == 'chr21'\")\n",
    "                         .assign(str_id = lambda x: [f\"{i}_{j}\" for i, j in zip(x[\"chr\"], x[\"start\"] + 1)]))\n",
    "\n",
    "df_repeats_grch38"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fb1b3da-36ac-4d43-8bd9-ffcf16cf132d",
   "metadata": {},
   "source": [
    "STR allele lengths for HG002 cell line that were called based on mappings of the maternal and paternal assemblies to GRCh38. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "42fcf1c3-1a4e-4e0b-8844-e14849633dd8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_id</th>\n",
       "      <th>region_len_ref</th>\n",
       "      <th>region_len_mat</th>\n",
       "      <th>region_len_pat</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>unit</th>\n",
       "      <th>period</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr21_5354757</td>\n",
       "      <td>12</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5354756</td>\n",
       "      <td>5354768</td>\n",
       "      <td>ATA</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr21_7205148</td>\n",
       "      <td>18</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>7205147</td>\n",
       "      <td>7205165</td>\n",
       "      <td>ATA</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr21_7227685</td>\n",
       "      <td>20</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>7227684</td>\n",
       "      <td>7227704</td>\n",
       "      <td>TTTATTTTTA</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr21_8987815</td>\n",
       "      <td>16</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>8987814</td>\n",
       "      <td>8987830</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr21_8987982</td>\n",
       "      <td>10</td>\n",
       "      <td>10.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>8987981</td>\n",
       "      <td>8987991</td>\n",
       "      <td>CGCCC</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19452</th>\n",
       "      <td>chr21_46688832</td>\n",
       "      <td>32</td>\n",
       "      <td>24.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>46688831</td>\n",
       "      <td>46688863</td>\n",
       "      <td>AC</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19453</th>\n",
       "      <td>chr21_46689712</td>\n",
       "      <td>12</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>46689711</td>\n",
       "      <td>46689723</td>\n",
       "      <td>ATTT</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19454</th>\n",
       "      <td>chr21_46696282</td>\n",
       "      <td>11</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>46696281</td>\n",
       "      <td>46696292</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19455</th>\n",
       "      <td>chr21_46697535</td>\n",
       "      <td>20</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>46697534</td>\n",
       "      <td>46697554</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19456</th>\n",
       "      <td>chr21_46699460</td>\n",
       "      <td>12</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>46699459</td>\n",
       "      <td>46699471</td>\n",
       "      <td>GGGC</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19457 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               str_id  region_len_ref  region_len_mat  region_len_pat  \\\n",
       "0       chr21_5354757              12            12.0            12.0   \n",
       "1       chr21_7205148              18            18.0            18.0   \n",
       "2       chr21_7227685              20            20.0            20.0   \n",
       "3       chr21_8987815              16            16.0            16.0   \n",
       "4       chr21_8987982              10            10.0            10.0   \n",
       "...               ...             ...             ...             ...   \n",
       "19452  chr21_46688832              32            24.0            32.0   \n",
       "19453  chr21_46689712              12            12.0            12.0   \n",
       "19454  chr21_46696282              11            11.0            11.0   \n",
       "19455  chr21_46697535              20            20.0            20.0   \n",
       "19456  chr21_46699460              12            12.0            12.0   \n",
       "\n",
       "          start       end        unit  period  \n",
       "0       5354756   5354768         ATA       3  \n",
       "1       7205147   7205165         ATA       3  \n",
       "2       7227684   7227704  TTTATTTTTA      10  \n",
       "3       8987814   8987830          CG       2  \n",
       "4       8987981   8987991       CGCCC       5  \n",
       "...         ...       ...         ...     ...  \n",
       "19452  46688831  46688863          AC       2  \n",
       "19453  46689711  46689723        ATTT       4  \n",
       "19454  46696281  46696292           T       1  \n",
       "19455  46697534  46697554           A       1  \n",
       "19456  46699459  46699471        GGGC       4  \n",
       "\n",
       "[19457 rows x 8 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_repeats_hg002 = pd.read_csv(\"../../data/HG002/variants/HG002_GRCh38_STR_lengths.csv\")\n",
    "df_repeats_hg002 = df_repeats_hg002[df_repeats_hg002[\"str_id\"].str.startswith(\"chr21\")]\n",
    "df_repeats_hg002 = df_repeats_hg002.merge(df_repeats_grch38[[\"str_id\", \"start\", \"end\", \"unit\", \"period\"]], on=\"str_id\", how=\"inner\")\n",
    "df_repeats_hg002"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94d98287-2bf8-4384-89da-d55e109cc929",
   "metadata": {},
   "source": [
    "Generate maternal and paternal haplotype specific data frames. For loci that are already the same length as in GRCh38, we don't need to change anything. So, select only loci that have different length from GRCh38 (but only differ by a multiple of the STR unit length). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b752a330-3c97-44e3-a632-4f3e0233f46a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((3824, 9), (3678, 9))"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mat_nonref = df_repeats_hg002.query(\"region_len_ref != region_len_mat and region_len_mat % period == 0\").assign(allele_len = lambda x: x[\"region_len_mat\"].convert_dtypes(int) // x[\"period\"])\n",
    "df_pat_nonref = df_repeats_hg002.query(\"region_len_ref != region_len_pat and region_len_pat % period == 0\").assign(allele_len = lambda x: x[\"region_len_pat\"].convert_dtypes(int) // x[\"period\"])\n",
    "df_mat_nonref.shape, df_pat_nonref.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "50e9e12f-ae94-42fa-86fe-b36115b8d719",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_id</th>\n",
       "      <th>region_len_ref</th>\n",
       "      <th>region_len_mat</th>\n",
       "      <th>region_len_pat</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>unit</th>\n",
       "      <th>period</th>\n",
       "      <th>allele_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr21_10382937</td>\n",
       "      <td>18</td>\n",
       "      <td>19.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>10382936</td>\n",
       "      <td>10382954</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr21_10386619</td>\n",
       "      <td>11</td>\n",
       "      <td>10.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>10386618</td>\n",
       "      <td>10386629</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chr21_10399238</td>\n",
       "      <td>21</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>10399237</td>\n",
       "      <td>10399258</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr21_10403964</td>\n",
       "      <td>35</td>\n",
       "      <td>40.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>10403963</td>\n",
       "      <td>10403998</td>\n",
       "      <td>TTTAT</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>chr21_10486318</td>\n",
       "      <td>24</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>10486317</td>\n",
       "      <td>10486341</td>\n",
       "      <td>AT</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19412</th>\n",
       "      <td>chr21_46632747</td>\n",
       "      <td>13</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>46632746</td>\n",
       "      <td>46632759</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19417</th>\n",
       "      <td>chr21_46639659</td>\n",
       "      <td>11</td>\n",
       "      <td>11.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>46639658</td>\n",
       "      <td>46639669</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19418</th>\n",
       "      <td>chr21_46640439</td>\n",
       "      <td>21</td>\n",
       "      <td>21.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>46640438</td>\n",
       "      <td>46640459</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19428</th>\n",
       "      <td>chr21_46645267</td>\n",
       "      <td>14</td>\n",
       "      <td>119.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>46645266</td>\n",
       "      <td>46645280</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19449</th>\n",
       "      <td>chr21_46682705</td>\n",
       "      <td>46</td>\n",
       "      <td>33.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>46682704</td>\n",
       "      <td>46682750</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3678 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               str_id  region_len_ref  region_len_mat  region_len_pat  \\\n",
       "16     chr21_10382937              18            19.0            19.0   \n",
       "21     chr21_10386619              11            10.0            10.0   \n",
       "31     chr21_10399238              21            20.0            20.0   \n",
       "36     chr21_10403964              35            40.0            40.0   \n",
       "39     chr21_10486318              24            20.0            20.0   \n",
       "...               ...             ...             ...             ...   \n",
       "19412  chr21_46632747              13            14.0            14.0   \n",
       "19417  chr21_46639659              11            11.0            12.0   \n",
       "19418  chr21_46640439              21            21.0            20.0   \n",
       "19428  chr21_46645267              14           119.0           120.0   \n",
       "19449  chr21_46682705              46            33.0            51.0   \n",
       "\n",
       "          start       end   unit  period  allele_len  \n",
       "16     10382936  10382954      A       1          19  \n",
       "21     10386618  10386629      T       1          10  \n",
       "31     10399237  10399258      T       1          20  \n",
       "36     10403963  10403998  TTTAT       5           8  \n",
       "39     10486317  10486341     AT       2          10  \n",
       "...         ...       ...    ...     ...         ...  \n",
       "19412  46632746  46632759      A       1          14  \n",
       "19417  46639658  46639669      T       1          12  \n",
       "19418  46640438  46640459      T       1          20  \n",
       "19428  46645266  46645280      A       1         120  \n",
       "19449  46682704  46682750      A       1          51  \n",
       "\n",
       "[3678 rows x 9 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pat_nonref\n",
    "# df_pat_nonref.query(\"region_len_pat != region_len_mat and period != 1\").head(50)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c975a684-eb73-403d-9d4e-9a19b7db7586",
   "metadata": {},
   "source": [
    "## Generating new chromosome sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9aa2ccec-fd73-488a-a37c-2d66076aad18",
   "metadata": {},
   "source": [
    "Load chr21 index through pysam. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "9abe8e2c-80cc-47fc-9750-90aa62661ee0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['chr21'] [46709983]\n"
     ]
    }
   ],
   "source": [
    "chr21 = pysam.FastaFile(\"../../data/simulated_reads/GRCh38_chr21.fa.gz\")\n",
    "print(chr21.references, chr21.lengths)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7415bafa-10a4-4865-92f0-148f1ac93047",
   "metadata": {},
   "source": [
    "### Paternal haplotye\n",
    "We do the paternal haplotype first. Based on the STR start and end positions, we determine the regions constituting the 'gaps' between STR loci, and the coordinates of the STR loci themselves. To generate a new chr21 representation, we extract the gaps directly from GRCh38's chr21 sequence. The STR loci (that should all be different from GRCh38 at this point), we will generate on the fly and fill them in between the gaps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "3a48f90a-d939-4fb6-a1d2-72ee477e2388",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3678 3678\n"
     ]
    }
   ],
   "source": [
    "strs_pat = list(zip(df_pat_nonref[\"unit\"], df_pat_nonref[\"allele_len\"]))\n",
    "\n",
    "gaps_pat = list(zip(df_pat_nonref[\"end\"].shift(1, fill_value=0), df_pat_nonref[\"start\"]))\n",
    "tail_pat = (df_pat_nonref.at[df_pat_nonref.index[-1], \"end\"], None) #make sure to keep track of the section of the chromosome after the last STR\n",
    "\n",
    "print(len(gaps_pat), len(strs_pat))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "1939b728-636b-4146-a84b-d65a89e3e9fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "pat_seq = \"\"\n",
    "for (gap_start, gap_end), (unit, period) in zip(gaps_pat, strs_pat):\n",
    "    pat_seq += chr21.fetch(\"chr21\", gap_start, gap_end)\n",
    "    pat_seq += unit * period\n",
    "pat_seq += chr21.fetch(\"chr21\", tail_pat[0], tail_pat[1])\n",
    "\n",
    "SeqIO.write(\n",
    "    SeqRecord(Seq(pat_seq), id=\"HG002_chr21_paternal\", description=\"Sequence based on GRCh38 chr21, with only STR sequences adapted to represent the paternal haplotype of the HG002 cell line\"), \n",
    "    \"../../data/simulated_reads/HG002_chr21_paternal.fa\", \n",
    "    format=\"fasta\"\n",
    ")\n",
    "\n",
    "new_seq = None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffa9cab6-701c-4233-bc07-96311aa5195a",
   "metadata": {},
   "source": [
    "### Maternal haplotype\n",
    "Then, do exactly the same for the maternal haplotype."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "2ed5d864-1ad7-4327-823e-aa0fdbe75f4b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3824 3824\n"
     ]
    }
   ],
   "source": [
    "strs_mat = list(zip(df_mat_nonref[\"unit\"], df_mat_nonref[\"allele_len\"]))\n",
    "\n",
    "gaps_mat = list(zip(df_mat_nonref[\"end\"].shift(1, fill_value=0), df_mat_nonref[\"start\"]))\n",
    "tail_mat = (df_mat_nonref.at[df_mat_nonref.index[-1], \"end\"], None)\n",
    "\n",
    "print(len(gaps_mat), len(strs_mat))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "3f5cd477-32bf-47fc-a6d5-dcc3d03ea51e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_seq = \"\"\n",
    "for (gap_start, gap_end), (unit, period) in zip(gaps_mat, strs_mat):\n",
    "    mat_seq += chr21.fetch(\"chr21\", gap_start, gap_end)\n",
    "    mat_seq += unit * period\n",
    "mat_seq += chr21.fetch(\"chr21\", tail_mat[0], tail_mat[1])\n",
    "\n",
    "SeqIO.write(\n",
    "    SeqRecord(Seq(mat_seq), id=\"HG002_chr21_maternal\", description=\"Sequence based on GRCh38 chr21, with only STR sequences adapted to represent the maternal haplotype of the HG002 cell line\"), \n",
    "    \"../../data/simulated_reads/HG002_chr21_maternal.fa\", \n",
    "    format=\"fasta\"\n",
    ")\n",
    "\n",
    "new_seq = None"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
