{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12cd304d-31c1-492e-9f2a-692e9c6d7274",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "# SRC_DIR environment variable should be the absolute path to the 'multicopy-STR-genotyping' directory\n",
    "sys.path.append(os.environ[\"SRC_DIR\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae5384da-70e5-42f1-81ef-f08ffcd3cf2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from multicopy_STR_genotyping import file_io\n",
    "\n",
    "sns.set_context(\"poster\")\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d66b0601-76e6-4d2b-b1f1-d0aea2f1ea6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf = ... # str: path to input vcf file                      \n",
    "vcf_format = ... # str: which tool was used to genotype\n",
    "sample = \"HG002\" # str: name of sample in vcf to analyse\n",
    "set_cn = ... # bool: set CN of loci manually (based on chromosome name)\n",
    "gt_in_bp = ... # bool: does the tool report genotypes in bp (True) or in number of repeat units (False)\n",
    "strs_covered_by_haplotypes = \"../../../data/HG002/regions/hg38_ver13_0boe_mononucleotides_union.bed\" # str: path to bed with loci that are covered by the haplotypes provided by Q100\n",
    "haplotype_str_length = \"../../../data/HG002/variants/HG002_GRCh38_STR_lengths.csv\" # str: path to csv with STR allele lengths based on Q100 haplotypes\n",
    "comparison_file_out = None # None or str: path to write csv with comparison of STR allele lengths from vcf to Q100 STR allele lengths\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d168c683-ab62-42a2-aff0-c6c989fb02d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_cn_manually(df):\n",
    "    autosomes = [f\"chr{i}\" for i in range(1, 23)]\n",
    "    df = df.assign(\n",
    "        copy_number = lambda x: [2 if i in autosomes else 1 for i in x[\"chr\"]]\n",
    "    )\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "855a158a-0d2e-4968-91b2-5a329c7c8d2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_repeats, df_hg002 = file_io.dfs_from_vcf(filename=vcf, samples=[sample], vcf_format=vcf_format)\n",
    "print(df_repeats.shape, df_hg002.shape)\n",
    "\n",
    "df_covered_repeats = pd.read_csv(strs_covered_by_haplotypes, sep=\"\\t\", names=[\"chr\", \"start\", \"end\", \"period\", \"unit\"])\n",
    "df_covered_repeats = df_covered_repeats.assign(\n",
    "            str_id = np.array([f\"{chrom}_{start + 1}\" for chrom, start in zip(df_covered_repeats.chr, df_covered_repeats.start)])\n",
    ")\n",
    "\n",
    "df_covered_repeats = df_repeats[df_repeats[\"str_id\"].isin(df_covered_repeats[\"str_id\"])]\n",
    "\n",
    "df_hg002 = df_hg002.merge(\n",
    "    df_covered_repeats[[\"str_id\", \"chr\", \"start\", \"end\", \"period\", \"unit\", \"ref\"]], on=\"str_id\", how=\"inner\"\n",
    ")\n",
    "\n",
    "if set_cn:\n",
    "    df_hg002 = set_cn_manually(df_hg002)\n",
    "    \n",
    "df_hg002 = df_hg002.dropna().reset_index(drop=True)\n",
    "\n",
    "print(df_covered_repeats.shape, df_hg002.shape)\n",
    "\n",
    "df_hg002\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32f504da-3262-4d81-ba22-8d267b921546",
   "metadata": {},
   "outputs": [],
   "source": [
    "if gt_in_bp:\n",
    "    df_hg002 = df_hg002.assign(\n",
    "        illumina_region_len = lambda x: [sorted(gt) for gt in x[\"genotype\"]],\n",
    "        depth = lambda x: [sum(i.values()) for i in x[\"frequencies\"]],\n",
    "    )\n",
    "else:\n",
    "    df_hg002 = df_hg002.assign(\n",
    "        illumina_region_len = lambda x: [sorted([k * j for k in i]) for i, j in zip(x[\"genotype\"], x[\"period\"])],\n",
    "        depth = lambda x: [sum(i.values()) for i in x[\"frequencies\"]],\n",
    "    )\n",
    "    \n",
    "df_hg002"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b40f6e52-73ae-4131-8d27-da8502f0c066",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_strs_haplotypes = pd.read_csv(haplotype_str_length).assign(chr = lambda x: [i[0] for i in x[\"str_id\"].str.split(\"_\")])\n",
    "df_strs_haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7fb63b2-8c61-45e6-a6c9-9ec20c9436a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "autosomes = \"chr != 'chrX' and chr != 'chrY'\"\n",
    "X = \"chr == 'chrX'\"\n",
    "Y = \"chr == 'chrY'\"\n",
    "\n",
    "df_strs_haplotypes[\"haplo_region_len\"] = [[] for i in range(len(df_strs_haplotypes))]\n",
    "\n",
    "df_strs_haplotypes.loc[df_strs_haplotypes.eval(autosomes), \"haplo_region_len\"] = pd.Series([\n",
    "    sorted([np.int64(i), np.int64(j)]) for i, j in zip(\n",
    "        df_strs_haplotypes.loc[df_strs_haplotypes.eval(autosomes), \"region_len_mat\"], \n",
    "        df_strs_haplotypes.loc[df_strs_haplotypes.eval(autosomes), \"region_len_pat\"],)])\n",
    "\n",
    "for i, j in zip(\n",
    "        df_strs_haplotypes.loc[df_strs_haplotypes.eval(X), \"haplo_region_len\"],\n",
    "        df_strs_haplotypes.loc[df_strs_haplotypes.eval(X), \"region_len_mat\"]):\n",
    "    i.append(np.int64(j))\n",
    "    \n",
    "for i, j in zip(\n",
    "        df_strs_haplotypes.loc[df_strs_haplotypes.eval(Y), \"haplo_region_len\"],\n",
    "        df_strs_haplotypes.loc[df_strs_haplotypes.eval(Y), \"region_len_pat\"]):\n",
    "    i.append(np.int64(j))\n",
    "    \n",
    "df_strs_haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5802936d-93d0-4b9b-b345-0b9d12d845a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_comparison = (\n",
    "    df_strs_haplotypes[[\"str_id\", \"haplo_region_len\"]]\n",
    "        .merge(df_hg002[[\"str_id\", \"illumina_region_len\", \"depth\", \"copy_number\", \"period\"]], on=\"str_id\", how=\"inner\")\n",
    "        .assign(\n",
    "            match = lambda x: x[\"haplo_region_len\"] == x[\"illumina_region_len\"],\n",
    "        )\n",
    ")\n",
    "\n",
    "df_comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5051a0a-ad35-4664-b1a0-03c207bb7f14",
   "metadata": {},
   "outputs": [],
   "source": [
    "if comparison_file_out:\n",
    "    df_comparison.to_csv(comparison_file_out, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1be5a7b-8489-4d5c-ad31-baa78240d46a",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_wrong = df_comparison.loc[~df_comparison[\"match\"]].shape[0]\n",
    "n_total = df_comparison.shape[0]\n",
    "\n",
    "print(f\"n correct: {n_total - n_wrong}, n wrong: {n_wrong}, n total: {n_total}\")\n",
    "print(\"Overall accuracy:\", (1 - (n_wrong / n_total)) * 100)\n",
    "\n",
    "df_comparison.loc[~df_comparison[\"match\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8c3002c",
   "metadata": {},
   "outputs": [],
   "source": [
    "min_depth = 15\n",
    "max_depth = 85\n",
    "query = f\"depth / copy_number >= {min_depth} and depth / copy_number <= {max_depth}\"\n",
    "\n",
    "n_wrong = df_comparison.query(query).loc[~df_comparison.query(query)[\"match\"]].shape[0]\n",
    "n_total = df_comparison.query(query).shape[0]\n",
    "\n",
    "print(f\"n correct: {n_total - n_wrong}, n wrong: {n_wrong}, n total: {n_total}\")\n",
    "print(\"Overall accuracy:\", (1 - (n_wrong / n_total)) * 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3dfee9f-e4aa-4948-8235-f762e044119e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_plot = pd.DataFrame(\n",
    "    np.concatenate([np.repeat([1, 2, 3, 4, 5, 6], 7).reshape(-1, 1), np.tile([1, 5, 5, 10, 10, 15, 15, 20, 20, 25, 25, 30, 30, 1e6], 6).reshape(-1, 2)], axis=1),\n",
    "    columns=[\"period\", \"depth_lower\", \"depth_upper\"]\n",
    ")\n",
    "\n",
    "accuracy = []\n",
    "n = []\n",
    "for row in df_plot.to_dict(orient=\"records\"):\n",
    "    selected_loci = (\n",
    "        df_comparison\n",
    "            .assign(depth = lambda x: x[\"depth\"] / x[\"copy_number\"])\n",
    "            .query(f\"depth >= {row['depth_lower']} and depth < {row['depth_upper']} and period == {row['period']}\")[\"str_id\"]\n",
    "    )\n",
    "    df_comparison_filt = df_comparison.loc[df_comparison[\"str_id\"].isin(selected_loci)]\n",
    "    n.append(df_comparison_filt.shape[0])\n",
    "    accuracy.append((1 - (df_comparison_filt.loc[~df_comparison_filt[\"match\"]].shape[0] / df_comparison_filt.shape[0])) * 100)\n",
    "\n",
    "df_plot[\"n\"] = n\n",
    "df_plot[\"accuracy\"] = accuracy\n",
    "df_plot[\"depth\"] = np.tile([\"1-5\", \"5-10\", \"10-15\", \"15-20\", \"20-25\", \"25-30\", \"30-\"], 6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "799024a0-855b-499a-8249-48b3aa33e4ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "for name, data in df_plot.groupby(\"depth\"):\n",
    "    print(name, np.average(data[\"accuracy\"], weights=data[\"n\"]), data[\"n\"].sum())\n",
    "print()\n",
    "for name, data in df_plot.groupby(\"period\"):\n",
    "    print(name, np.average(data[\"accuracy\"], weights=data[\"n\"]), data[\"n\"].sum())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c70fd29f-5a96-4568-baf7-e6a9852feb2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(10, 8))\n",
    "\n",
    "ax = sns.barplot(\n",
    "    df_plot,\n",
    "    x=\"period\",\n",
    "    y=\"accuracy\",\n",
    "    hue=\"depth\",\n",
    "    palette=\"colorblind\"\n",
    ")\n",
    "\n",
    "ax.set(\n",
    "    xlabel=\"Period\",\n",
    "    ylabel = \"Accuracy\"\n",
    ")\n",
    "\n",
    "_ = ax.get_legend().set(\n",
    "    title=\"Depth / CN\",\n",
    "    bbox_to_anchor = (1, 1)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0767684-6c05-41a5-9d1b-b9067e446168",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import ndimage\n",
    "\n",
    "df_plot_density = (\n",
    "    df_comparison\n",
    "        .query(\"period >= 1 and period <= 6\")\n",
    "        .assign(depth = lambda x: x[\"depth\"] / x[\"copy_number\"])[[\"match\", \"period\", \"depth\"]]\n",
    ")\n",
    "\n",
    "df_plot_smooth = (\n",
    "    df_plot_density        \n",
    "        .drop(\"period\", axis=1)\n",
    "        .groupby([\"depth\"], observed=True, as_index=False).agg(\n",
    "            accuracy = (\"match\", lambda x: x.sum() / len(x)), \n",
    "            n_obs=(\"match\", \"count\"))\n",
    "        # .query(f\"n_obs >= {min_obs}\")\n",
    "        \n",
    ")\n",
    "\n",
    "smooth = []\n",
    "smooth = ndimage.gaussian_filter1d(df_plot_smooth[\"accuracy\"], 3, mode=\"reflect\")\n",
    "df_plot_smooth = df_plot_smooth.assign(smooth = smooth)\n",
    "\n",
    "\n",
    "df_plot_smooth_period = (\n",
    "    df_plot_density\n",
    "        .groupby([\"period\", \"depth\"], observed=True, as_index=False).agg(\n",
    "            accuracy = (\"match\", lambda x: x.sum() / len(x)), \n",
    "            n_obs=(\"match\", \"count\"))\n",
    "        # .query(f\"n_obs >= {min_obs}\")\n",
    "        \n",
    ")\n",
    "\n",
    "smooth = []\n",
    "for group, data in df_plot_smooth_period.groupby([\"period\"]):\n",
    "    current_smooth = ndimage.gaussian_filter1d(data[\"accuracy\"], 3, mode=\"reflect\")\n",
    "    smooth.append(current_smooth)\n",
    "df_plot_smooth_period = df_plot_smooth_period.assign(smooth = np.concatenate(smooth))\n",
    "\n",
    "df_plot_smooth_period"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9093aee2-cac1-4433-a913-790e7a2c9b8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(dpi=125)\n",
    "\n",
    "ax = sns.lineplot(\n",
    "    df_plot_smooth,\n",
    "    x = \"depth\",\n",
    "    y = \"smooth\",\n",
    "    color=sns.color_palette(\"colorblind\")[0],\n",
    "    # linewidth=2,\n",
    "    legend=False\n",
    ")\n",
    "\n",
    "ax = sns.scatterplot(\n",
    "    df_plot_smooth,\n",
    "    x = \"depth\",\n",
    "    y = \"accuracy\",\n",
    "    color=sns.color_palette(\"colorblind\")[0],\n",
    "    s=25,\n",
    "    ax = ax,\n",
    "    legend=False\n",
    ")\n",
    "\n",
    "ax.set(\n",
    "    xlabel = \"Sequencing depth / CN\",\n",
    "    ylabel = \"Accuracy\",\n",
    "    ylim = (0, 1.05),\n",
    "    xlim = (0, 55)\n",
    ")\n",
    "\n",
    "ax2 = ax.twinx()\n",
    "ax2 = sns.histplot(\n",
    "    df_plot_density,\n",
    "    x = \"depth\",\n",
    "    discrete=True,\n",
    "    stat=\"proportion\",\n",
    "    color=\"grey\",\n",
    "    ax=ax2,\n",
    ")\n",
    "ax2.set(\n",
    "    ylabel=\"Proportion of STR loci\",\n",
    "    ylim=(0, 0.07),\n",
    ")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95431594-6d2c-4935-af32-16a0a1b7582e",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(3, 2, figsize=(20, 20))\n",
    "\n",
    "for period, ax in zip(range(1, 7), axes.ravel()):\n",
    "    df_subplot = (df_plot_smooth_period\n",
    "                  .query(f\"period == {period}\"))\n",
    "\n",
    "    ax = sns.lineplot(\n",
    "        df_subplot,\n",
    "        x = \"depth\",\n",
    "        y = \"smooth\",\n",
    "        color=sns.color_palette(\"colorblind\")[0],\n",
    "        linewidth=5,\n",
    "        ax = ax,\n",
    "    )\n",
    "\n",
    "    ax = sns.scatterplot(\n",
    "        df_subplot,\n",
    "        x = \"depth\",\n",
    "        y = \"accuracy\",\n",
    "        color=sns.color_palette(\"colorblind\")[0],\n",
    "        s=25,\n",
    "        ax = ax,\n",
    "    )\n",
    "\n",
    "    ax.set(\n",
    "        title = f\"Period {period}\",\n",
    "        xlabel = \"Sequencing depth / CN\",\n",
    "        ylabel = \"Accuracy\",\n",
    "        ylim = (0, 1.05),\n",
    "        xlim = (0, 55)\n",
    "    )\n",
    "    \n",
    "    ax2 = ax.twinx()\n",
    "    ax2 = sns.histplot(\n",
    "        df_plot_density.query(f\"period == {period}\"),\n",
    "        x = \"depth\",\n",
    "        discrete=True,\n",
    "        stat=\"proportion\",\n",
    "        color=\"grey\",\n",
    "        ax=ax2,\n",
    "    )\n",
    "    \n",
    "    ax2.set(\n",
    "        ylabel=\"Proportion of STR loci\",\n",
    "        ylim=(0, 0.08),\n",
    "    )\n",
    "\n",
    "plt.tight_layout()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
