{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a3cc9f5d-7801-4a40-a86d-07495e1fc4dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "# SRC_DIR should be the absolute path to the 'multicopy-STR-genotyping' directory\n",
    "sys.path.append(os.environ[\"SRC_DIR\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "71cb5ffd-9217-4cc7-8fbe-b8cc4fc6c164",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from multicopy_STR_genotyping import file_io"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "97cc4c77-f56c-46c0-b5ca-fd81f326fb95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3438809, 5)\n",
      "(3327344, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>40639</td>\n",
       "      <td>40639</td>\n",
       "      <td>-</td>\n",
       "      <td>tttt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>41052</td>\n",
       "      <td>41053</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>41217</td>\n",
       "      <td>41218</td>\n",
       "      <td>t</td>\n",
       "      <td>a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>41255</td>\n",
       "      <td>41256</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>41387</td>\n",
       "      <td>41388</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  start    end REF   ALT\n",
       "0  chr1  40639  40639   -  tttt\n",
       "1  chr1  41052  41053   c     t\n",
       "2  chr1  41217  41218   t     a\n",
       "3  chr1  41255  41256   c     t\n",
       "4  chr1  41387  41388   a     g"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# chr, start, end, query depth, mapping quality, REF allele, ALT allele, query name, query start, end and the query orientation\n",
    "haplotype_maternal = \"../../data/HG002/alignments/HG002.maternal.f1_assembly_v2_genbank_sort_var.txt\"\n",
    "haplotype_paternal = \"../../data/HG002/alignments/HG002.paternal.f1_assembly_v2_genbank_sort_var.txt\"\n",
    "\n",
    "df_haplo_mat = file_io.load_haplotype_vars(haplotype_maternal)\n",
    "print(df_haplo_mat.shape)\n",
    "\n",
    "df_haplo_pat = file_io.load_haplotype_vars(haplotype_paternal)\n",
    "print(df_haplo_pat.shape)\n",
    "\n",
    "\n",
    "df_haplo_pat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "532a7944-8f07-4a08-ba84-54dc5c4ba8c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(686236, 5)\n",
      "(663895, 5)\n"
     ]
    }
   ],
   "source": [
    "df_haplo_mat_indel = df_haplo_mat.query(\"REF == '-' or ALT == '-'\")\n",
    "print(df_haplo_mat_indel.shape)\n",
    "\n",
    "df_haplo_pat_indel = df_haplo_pat.query(\"REF == '-' or ALT == '-'\")\n",
    "print(df_haplo_pat_indel.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "8aad0237-34d4-46c8-b8e3-5db0a5947b03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>str_id</th>\n",
       "      <th>copy_number</th>\n",
       "      <th>frequencies</th>\n",
       "      <th>genotype</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>period</th>\n",
       "      <th>unit</th>\n",
       "      <th>ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_588068</td>\n",
       "      <td>2</td>\n",
       "      <td>{4: 22}</td>\n",
       "      <td>[4, 4]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>588068</td>\n",
       "      <td>588079</td>\n",
       "      <td>3</td>\n",
       "      <td>CCT</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_589245</td>\n",
       "      <td>2</td>\n",
       "      <td>{6: 9}</td>\n",
       "      <td>[6, 6]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>589245</td>\n",
       "      <td>589256</td>\n",
       "      <td>2</td>\n",
       "      <td>TC</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_590659</td>\n",
       "      <td>2</td>\n",
       "      <td>{3: 60}</td>\n",
       "      <td>[3, 3]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>590659</td>\n",
       "      <td>590670</td>\n",
       "      <td>4</td>\n",
       "      <td>AAAT</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_590969</td>\n",
       "      <td>2</td>\n",
       "      <td>{4: 42}</td>\n",
       "      <td>[4, 4]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>590969</td>\n",
       "      <td>590984</td>\n",
       "      <td>4</td>\n",
       "      <td>AAAC</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_594083</td>\n",
       "      <td>2</td>\n",
       "      <td>{4: 21}</td>\n",
       "      <td>[4, 4]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>594083</td>\n",
       "      <td>594094</td>\n",
       "      <td>3</td>\n",
       "      <td>TCC</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>813889</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chrY_26625608</td>\n",
       "      <td>1</td>\n",
       "      <td>{2: 43}</td>\n",
       "      <td>[2]</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26625608</td>\n",
       "      <td>26625621</td>\n",
       "      <td>7</td>\n",
       "      <td>TAAGTAT</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>813890</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chrY_26627699</td>\n",
       "      <td>1</td>\n",
       "      <td>{4: 23}</td>\n",
       "      <td>[4]</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26627699</td>\n",
       "      <td>26627710</td>\n",
       "      <td>3</td>\n",
       "      <td>GAA</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>813891</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chrY_26628607</td>\n",
       "      <td>1</td>\n",
       "      <td>{3: 31}</td>\n",
       "      <td>[3]</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26628607</td>\n",
       "      <td>26628621</td>\n",
       "      <td>5</td>\n",
       "      <td>ATTTT</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>813892</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chrY_26632688</td>\n",
       "      <td>1</td>\n",
       "      <td>{6: 9}</td>\n",
       "      <td>[6]</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26632688</td>\n",
       "      <td>26632705</td>\n",
       "      <td>3</td>\n",
       "      <td>TTC</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>813893</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chrY_26637436</td>\n",
       "      <td>1</td>\n",
       "      <td>{4: 33}</td>\n",
       "      <td>[4]</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26637436</td>\n",
       "      <td>26637447</td>\n",
       "      <td>3</td>\n",
       "      <td>TTA</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>813894 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    sample         str_id  copy_number frequencies genotype  \\\n",
       "0       HG002.GRCh38.2x250    chr1_588068            2     {4: 22}   [4, 4]   \n",
       "1       HG002.GRCh38.2x250    chr1_589245            2      {6: 9}   [6, 6]   \n",
       "2       HG002.GRCh38.2x250    chr1_590659            2     {3: 60}   [3, 3]   \n",
       "3       HG002.GRCh38.2x250    chr1_590969            2     {4: 42}   [4, 4]   \n",
       "4       HG002.GRCh38.2x250    chr1_594083            2     {4: 21}   [4, 4]   \n",
       "...                    ...            ...          ...         ...      ...   \n",
       "813889  HG002.GRCh38.2x250  chrY_26625608            1     {2: 43}      [2]   \n",
       "813890  HG002.GRCh38.2x250  chrY_26627699            1     {4: 23}      [4]   \n",
       "813891  HG002.GRCh38.2x250  chrY_26628607            1     {3: 31}      [3]   \n",
       "813892  HG002.GRCh38.2x250  chrY_26632688            1      {6: 9}      [6]   \n",
       "813893  HG002.GRCh38.2x250  chrY_26637436            1     {4: 33}      [4]   \n",
       "\n",
       "         chr     start       end  period     unit  ref  \n",
       "0       chr1    588068    588079       3      CCT    4  \n",
       "1       chr1    589245    589256       2       TC    6  \n",
       "2       chr1    590659    590670       4     AAAT    3  \n",
       "3       chr1    590969    590984       4     AAAC    4  \n",
       "4       chr1    594083    594094       3      TCC    4  \n",
       "...      ...       ...       ...     ...      ...  ...  \n",
       "813889  chrY  26625608  26625621       7  TAAGTAT    2  \n",
       "813890  chrY  26627699  26627710       3      GAA    4  \n",
       "813891  chrY  26628607  26628621       5    ATTTT    3  \n",
       "813892  chrY  26632688  26632705       3      TTC    6  \n",
       "813893  chrY  26637436  26637447       3      TTA    4  \n",
       "\n",
       "[813894 rows x 11 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_repeats, df_hg002 = file_io.dfs_from_vcf(file=\"../../data/HG002/variants/HG002.GRCh38.2x250.vcf\", samples=[\"HG002.GRCh38.2x250\"])\n",
    "\n",
    "df_covered_repeats = pd.read_csv(\"../../data/HG002/regions/hg38_ver13_union.bed\", sep=\"\\t\", names=[\"chr\", \"start\", \"end\", \"period\", \"unit\"])\n",
    "df_covered_repeats = df_covered_repeats.assign(\n",
    "            str_id = np.array([f\"{chrom}_{start + 1}\" for chrom, start in zip(df_covered_repeats.chr, df_covered_repeats.start)])\n",
    ")\n",
    "\n",
    "# df_covered_repeats = df_repeats.merge(df_covered_repeats[[\"str_id\", \"unit\"]], on=\"str_id\", how=\"inner\")/\n",
    "df_covered_repeats = df_repeats[df_repeats[\"str_id\"].isin(df_covered_repeats[\"str_id\"])]\n",
    "\n",
    "df_hg002 = df_hg002.merge(\n",
    "    df_covered_repeats[[\"str_id\", \"chr\", \"start\", \"end\", \"period\", \"unit\", \"ref\"]], on=\"str_id\", how=\"inner\"\n",
    ").reset_index(drop=True)\n",
    "\n",
    "df_hg002"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "f42ee3ab-0f3b-427d-b669-43b39d99980e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mask = []\n",
    "for idx, (ref, gt) in enumerate(zip(df_hg002[\"ref\"], df_hg002[\"genotype\"])):\n",
    "    try:\n",
    "        comp = any([allele != ref for allele in gt])\n",
    "        mask.append(comp)\n",
    "    except TypeError:\n",
    "        mask.append(False)\n",
    "\n",
    "mask = np.array(mask)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "14609ef3-6c1d-4f55-a31a-435ca640d8c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>str_id</th>\n",
       "      <th>copy_number</th>\n",
       "      <th>frequencies</th>\n",
       "      <th>genotype</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>period</th>\n",
       "      <th>unit</th>\n",
       "      <th>ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_650384</td>\n",
       "      <td>2</td>\n",
       "      <td>{18: 20, 20: 13}</td>\n",
       "      <td>[18, 20]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>650384</td>\n",
       "      <td>650423</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_744867</td>\n",
       "      <td>2</td>\n",
       "      <td>{7: 6, 8: 3}</td>\n",
       "      <td>[7, 8]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>744867</td>\n",
       "      <td>744882</td>\n",
       "      <td>2</td>\n",
       "      <td>TA</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_832737</td>\n",
       "      <td>2</td>\n",
       "      <td>{10: 38}</td>\n",
       "      <td>[10, 10]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>832737</td>\n",
       "      <td>832781</td>\n",
       "      <td>5</td>\n",
       "      <td>GTTTT</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_893802</td>\n",
       "      <td>2</td>\n",
       "      <td>{6: 12, 7: 16}</td>\n",
       "      <td>[6, 7]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>893802</td>\n",
       "      <td>893839</td>\n",
       "      <td>2</td>\n",
       "      <td>AT</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_931132</td>\n",
       "      <td>2</td>\n",
       "      <td>{3: 5, 4: 47}</td>\n",
       "      <td>[4, 4]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>931132</td>\n",
       "      <td>931143</td>\n",
       "      <td>4</td>\n",
       "      <td>CCCT</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_950660</td>\n",
       "      <td>2</td>\n",
       "      <td>{5: 71}</td>\n",
       "      <td>[5, 5]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>950660</td>\n",
       "      <td>950671</td>\n",
       "      <td>2</td>\n",
       "      <td>CA</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1005758</td>\n",
       "      <td>2</td>\n",
       "      <td>{1: 61}</td>\n",
       "      <td>[1, 1]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1005758</td>\n",
       "      <td>1005785</td>\n",
       "      <td>14</td>\n",
       "      <td>GCCCCCGCAGCAGT</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1010482</td>\n",
       "      <td>2</td>\n",
       "      <td>{3: 76}</td>\n",
       "      <td>[3, 3]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1010482</td>\n",
       "      <td>1010497</td>\n",
       "      <td>4</td>\n",
       "      <td>TTAT</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1293732</td>\n",
       "      <td>2</td>\n",
       "      <td>{3: 31, 4: 18}</td>\n",
       "      <td>[3, 4]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1293732</td>\n",
       "      <td>1293791</td>\n",
       "      <td>20</td>\n",
       "      <td>CGCCCCTGCCCTGGAGGCCC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1557496</td>\n",
       "      <td>2</td>\n",
       "      <td>{6: 29, 7: 36}</td>\n",
       "      <td>[6, 7]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1557496</td>\n",
       "      <td>1557509</td>\n",
       "      <td>2</td>\n",
       "      <td>AT</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1565106</td>\n",
       "      <td>2</td>\n",
       "      <td>{2: 27, 3: 16}</td>\n",
       "      <td>[2, 3]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1565106</td>\n",
       "      <td>1565156</td>\n",
       "      <td>17</td>\n",
       "      <td>AAGGCAGGCGGATCATG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1574033</td>\n",
       "      <td>2</td>\n",
       "      <td>{3: 20, 4: 18}</td>\n",
       "      <td>[3, 4]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1574033</td>\n",
       "      <td>1574044</td>\n",
       "      <td>3</td>\n",
       "      <td>AAG</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1585950</td>\n",
       "      <td>2</td>\n",
       "      <td>{17: 30, 18: 12, 19: 1}</td>\n",
       "      <td>[17, 18]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1585950</td>\n",
       "      <td>1585976</td>\n",
       "      <td>3</td>\n",
       "      <td>AAT</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1589161</td>\n",
       "      <td>2</td>\n",
       "      <td>{9: 38}</td>\n",
       "      <td>[9, 9]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1589161</td>\n",
       "      <td>1589200</td>\n",
       "      <td>4</td>\n",
       "      <td>TTTA</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1629847</td>\n",
       "      <td>2</td>\n",
       "      <td>{2: 8, 3: 6}</td>\n",
       "      <td>[2, 3]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1629847</td>\n",
       "      <td>1629867</td>\n",
       "      <td>7</td>\n",
       "      <td>CCCCCAG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1653826</td>\n",
       "      <td>2</td>\n",
       "      <td>{18: 1, 19: 11, 22: 18}</td>\n",
       "      <td>[19, 22]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1653826</td>\n",
       "      <td>1653863</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1752909</td>\n",
       "      <td>2</td>\n",
       "      <td>{9: 37, 10: 27}</td>\n",
       "      <td>[9, 10]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1752909</td>\n",
       "      <td>1752935</td>\n",
       "      <td>3</td>\n",
       "      <td>CCT</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1762760</td>\n",
       "      <td>2</td>\n",
       "      <td>{7: 28, 8: 35}</td>\n",
       "      <td>[7, 8]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1762760</td>\n",
       "      <td>1762791</td>\n",
       "      <td>4</td>\n",
       "      <td>AATG</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1776462</td>\n",
       "      <td>2</td>\n",
       "      <td>{2: 42, 3: 42}</td>\n",
       "      <td>[2, 3]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1776462</td>\n",
       "      <td>1776473</td>\n",
       "      <td>4</td>\n",
       "      <td>AAAC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>385</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1845825</td>\n",
       "      <td>2</td>\n",
       "      <td>{20: 1, 21: 38, 24: 1}</td>\n",
       "      <td>[21, 21]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1845825</td>\n",
       "      <td>1845872</td>\n",
       "      <td>2</td>\n",
       "      <td>AC</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1857306</td>\n",
       "      <td>2</td>\n",
       "      <td>{3: 33, 4: 32}</td>\n",
       "      <td>[3, 4]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1857306</td>\n",
       "      <td>1857317</td>\n",
       "      <td>4</td>\n",
       "      <td>GGCT</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1876493</td>\n",
       "      <td>2</td>\n",
       "      <td>{10: 2, 11: 30, 12: 34}</td>\n",
       "      <td>[11, 12]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1876493</td>\n",
       "      <td>1876514</td>\n",
       "      <td>2</td>\n",
       "      <td>GA</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1898649</td>\n",
       "      <td>2</td>\n",
       "      <td>{1: 43, 2: 41}</td>\n",
       "      <td>[1, 2]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1898649</td>\n",
       "      <td>1898686</td>\n",
       "      <td>19</td>\n",
       "      <td>ACACACATAAGTTGGATAT</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1904425</td>\n",
       "      <td>2</td>\n",
       "      <td>{8: 28, 9: 30}</td>\n",
       "      <td>[8, 9]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1904425</td>\n",
       "      <td>1904448</td>\n",
       "      <td>4</td>\n",
       "      <td>AAAT</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>HG002.GRCh38.2x250</td>\n",
       "      <td>chr1_1919189</td>\n",
       "      <td>2</td>\n",
       "      <td>{2: 25, 3: 30}</td>\n",
       "      <td>[2, 3]</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1919189</td>\n",
       "      <td>1919215</td>\n",
       "      <td>9</td>\n",
       "      <td>AGCGGCAGG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sample        str_id  copy_number              frequencies  \\\n",
       "16   HG002.GRCh38.2x250   chr1_650384            2         {18: 20, 20: 13}   \n",
       "36   HG002.GRCh38.2x250   chr1_744867            2             {7: 6, 8: 3}   \n",
       "65   HG002.GRCh38.2x250   chr1_832737            2                 {10: 38}   \n",
       "77   HG002.GRCh38.2x250   chr1_893802            2           {6: 12, 7: 16}   \n",
       "89   HG002.GRCh38.2x250   chr1_931132            2            {3: 5, 4: 47}   \n",
       "97   HG002.GRCh38.2x250   chr1_950660            2                  {5: 71}   \n",
       "119  HG002.GRCh38.2x250  chr1_1005758            2                  {1: 61}   \n",
       "121  HG002.GRCh38.2x250  chr1_1010482            2                  {3: 76}   \n",
       "199  HG002.GRCh38.2x250  chr1_1293732            2           {3: 31, 4: 18}   \n",
       "291  HG002.GRCh38.2x250  chr1_1557496            2           {6: 29, 7: 36}   \n",
       "294  HG002.GRCh38.2x250  chr1_1565106            2           {2: 27, 3: 16}   \n",
       "297  HG002.GRCh38.2x250  chr1_1574033            2           {3: 20, 4: 18}   \n",
       "301  HG002.GRCh38.2x250  chr1_1585950            2  {17: 30, 18: 12, 19: 1}   \n",
       "303  HG002.GRCh38.2x250  chr1_1589161            2                  {9: 38}   \n",
       "318  HG002.GRCh38.2x250  chr1_1629847            2             {2: 8, 3: 6}   \n",
       "326  HG002.GRCh38.2x250  chr1_1653826            2  {18: 1, 19: 11, 22: 18}   \n",
       "357  HG002.GRCh38.2x250  chr1_1752909            2          {9: 37, 10: 27}   \n",
       "360  HG002.GRCh38.2x250  chr1_1762760            2           {7: 28, 8: 35}   \n",
       "366  HG002.GRCh38.2x250  chr1_1776462            2           {2: 42, 3: 42}   \n",
       "385  HG002.GRCh38.2x250  chr1_1845825            2   {20: 1, 21: 38, 24: 1}   \n",
       "390  HG002.GRCh38.2x250  chr1_1857306            2           {3: 33, 4: 32}   \n",
       "399  HG002.GRCh38.2x250  chr1_1876493            2  {10: 2, 11: 30, 12: 34}   \n",
       "408  HG002.GRCh38.2x250  chr1_1898649            2           {1: 43, 2: 41}   \n",
       "412  HG002.GRCh38.2x250  chr1_1904425            2           {8: 28, 9: 30}   \n",
       "417  HG002.GRCh38.2x250  chr1_1919189            2           {2: 25, 3: 30}   \n",
       "\n",
       "     genotype   chr    start      end  period                  unit  ref  \n",
       "16   [18, 20]  chr1   650384   650423       2                    AC   20  \n",
       "36     [7, 8]  chr1   744867   744882       2                    TA    8  \n",
       "65   [10, 10]  chr1   832737   832781       5                 GTTTT    9  \n",
       "77     [6, 7]  chr1   893802   893839       2                    AT   19  \n",
       "89     [4, 4]  chr1   931132   931143       4                  CCCT    3  \n",
       "97     [5, 5]  chr1   950660   950671       2                    CA    6  \n",
       "119    [1, 1]  chr1  1005758  1005785      14        GCCCCCGCAGCAGT    2  \n",
       "121    [3, 3]  chr1  1010482  1010497       4                  TTAT    4  \n",
       "199    [3, 4]  chr1  1293732  1293791      20  CGCCCCTGCCCTGGAGGCCC    3  \n",
       "291    [6, 7]  chr1  1557496  1557509       2                    AT    7  \n",
       "294    [2, 3]  chr1  1565106  1565156      17     AAGGCAGGCGGATCATG    3  \n",
       "297    [3, 4]  chr1  1574033  1574044       3                   AAG    4  \n",
       "301  [17, 18]  chr1  1585950  1585976       3                   AAT    9  \n",
       "303    [9, 9]  chr1  1589161  1589200       4                  TTTA   10  \n",
       "318    [2, 3]  chr1  1629847  1629867       7               CCCCCAG    3  \n",
       "326  [19, 22]  chr1  1653826  1653863       2                    AC   19  \n",
       "357   [9, 10]  chr1  1752909  1752935       3                   CCT    9  \n",
       "360    [7, 8]  chr1  1762760  1762791       4                  AATG    8  \n",
       "366    [2, 3]  chr1  1776462  1776473       4                  AAAC    3  \n",
       "385  [21, 21]  chr1  1845825  1845872       2                    AC   24  \n",
       "390    [3, 4]  chr1  1857306  1857317       4                  GGCT    3  \n",
       "399  [11, 12]  chr1  1876493  1876514       2                    GA   11  \n",
       "408    [1, 2]  chr1  1898649  1898686      19   ACACACATAAGTTGGATAT    2  \n",
       "412    [8, 9]  chr1  1904425  1904448       4                  AAAT    6  \n",
       "417    [2, 3]  chr1  1919189  1919215       9             AGCGGCAGG    3  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_hg002[mask].head(25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "61590bef-af97-4137-a30a-6ab68fcb1bd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2644</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1904424</td>\n",
       "      <td>1904424</td>\n",
       "      <td>-</td>\n",
       "      <td>aaataaat</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr    start      end REF       ALT\n",
       "2644  chr1  1904424  1904424   -  aaataaat"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx = 412\n",
    "locus = df_hg002.iloc[idx]\n",
    "# flank = 100\n",
    "\n",
    "# df_haplo_pat.query(f\"chr == '{locus.chr}' and start >= {locus.start} and end <= {locus.start}\")\n",
    "df_haplo_pat.query(f\"chr == '{locus.chr}' and start <= {locus.end} and {locus.start - 1} <= end\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "6c643be9-9400-41b1-81df-603bb035d9c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1942</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1904424</td>\n",
       "      <td>1904424</td>\n",
       "      <td>-</td>\n",
       "      <td>aaataaataaat</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr    start      end REF           ALT\n",
       "1942  chr1  1904424  1904424   -  aaataaataaat"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_haplo_mat.query(f\"chr == '{locus.chr}' and start <= {locus.end} and {locus.start} <= end\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
