{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a94c17b4-44c2-42bb-852c-00f64ef5826e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "# SRC_DIR environment variable should be the absolute path to the 'multicopy-STR-genotyping' directory\n",
    "sys.path.append(os.environ[\"SRC_DIR\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "98a3c657-bf66-40ab-80e8-059477b72c02",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from multicopy_STR_genotyping import file_io\n",
    "\n",
    "sns.set_context(\"poster\")\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "aac57752-76d9-489a-92cd-7f97919dd655",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3438809, 5)\n",
      "(3327344, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>40639</td>\n",
       "      <td>40639</td>\n",
       "      <td>-</td>\n",
       "      <td>tttt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>41052</td>\n",
       "      <td>41053</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>41217</td>\n",
       "      <td>41218</td>\n",
       "      <td>t</td>\n",
       "      <td>a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>41255</td>\n",
       "      <td>41256</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>41387</td>\n",
       "      <td>41388</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  start    end REF   ALT\n",
       "0  chr1  40639  40639   -  tttt\n",
       "1  chr1  41052  41053   c     t\n",
       "2  chr1  41217  41218   t     a\n",
       "3  chr1  41255  41256   c     t\n",
       "4  chr1  41387  41388   a     g"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# chr, start, end, query depth, mapping quality, REF allele, ALT allele, query name, query start, end and the query orientation\n",
    "haplotype_maternal = \"../../data/HG002/alignments/HG002.maternal.f1_assembly_v2_genbank_sort_var.txt\"\n",
    "haplotype_paternal = \"../../data/HG002/alignments/HG002.paternal.f1_assembly_v2_genbank_sort_var.txt\"\n",
    "\n",
    "df_haplo_mat = file_io.load_haplotype_vars(haplotype_maternal)\n",
    "print(df_haplo_mat.shape)\n",
    "\n",
    "df_haplo_pat = file_io.load_haplotype_vars(haplotype_paternal)\n",
    "print(df_haplo_pat.shape)\n",
    "\n",
    "\n",
    "df_haplo_pat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4b6584e7-cdd2-4ec0-9fed-e27389a303a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(686236, 5)\n",
      "(663895, 5)\n"
     ]
    }
   ],
   "source": [
    "df_haplo_mat_indel = df_haplo_mat.query(\"REF == '-' or ALT == '-'\")\n",
    "print(df_haplo_mat_indel.shape)\n",
    "\n",
    "df_haplo_pat_indel = df_haplo_pat.query(\"REF == '-' or ALT == '-'\")\n",
    "print(df_haplo_pat_indel.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "904134f2-aee6-4fa4-a805-9b5ce8a49ff7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>period</th>\n",
       "      <th>unit</th>\n",
       "      <th>str_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>588068</td>\n",
       "      <td>588079</td>\n",
       "      <td>3</td>\n",
       "      <td>CCT</td>\n",
       "      <td>chr1_588068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>589245</td>\n",
       "      <td>589256</td>\n",
       "      <td>2</td>\n",
       "      <td>TC</td>\n",
       "      <td>chr1_589245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>590659</td>\n",
       "      <td>590670</td>\n",
       "      <td>4</td>\n",
       "      <td>AAAT</td>\n",
       "      <td>chr1_590659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>590969</td>\n",
       "      <td>590984</td>\n",
       "      <td>4</td>\n",
       "      <td>AAAC</td>\n",
       "      <td>chr1_590969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>591734</td>\n",
       "      <td>591751</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_591734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695860</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57188868</td>\n",
       "      <td>57188878</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>chrY_57188868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695861</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57200838</td>\n",
       "      <td>57200851</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>chrY_57200838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695862</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201000</td>\n",
       "      <td>57201019</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>chrY_57201000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695863</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57202381</td>\n",
       "      <td>57202403</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>chrY_57202381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695864</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57202848</td>\n",
       "      <td>57202873</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>chrY_57202848</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1695865 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr     start       end  period  unit         str_id\n",
       "0        chr1    588068    588079       3   CCT    chr1_588068\n",
       "1        chr1    589245    589256       2    TC    chr1_589245\n",
       "2        chr1    590659    590670       4  AAAT    chr1_590659\n",
       "3        chr1    590969    590984       4  AAAC    chr1_590969\n",
       "4        chr1    591734    591751       1     A    chr1_591734\n",
       "...       ...       ...       ...     ...   ...            ...\n",
       "1695860  chrY  57188868  57188878       1     A  chrY_57188868\n",
       "1695861  chrY  57200838  57200851       1     A  chrY_57200838\n",
       "1695862  chrY  57201000  57201019       1     A  chrY_57201000\n",
       "1695863  chrY  57202381  57202403       1     A  chrY_57202381\n",
       "1695864  chrY  57202848  57202873       1     A  chrY_57202848\n",
       "\n",
       "[1695865 rows x 6 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_covered_repeats = (\n",
    "    pd.read_csv(\n",
    "        \"../../data/HG002/regions/hg38_ver13_0boe_mononucleotides_union.bed\", \n",
    "        sep=\"\\t\", \n",
    "        names=[\"chr\", \"start\", \"end\", \"period\", \"unit\"])\n",
    "    .assign(\n",
    "        start = lambda x: x[\"start\"] + 1, # convert coordinates from 0-based (BED) to 1-based (VCF, PAF)\n",
    "    ).assign(\n",
    "        str_id = lambda x: [f\"{i}_{j}\" for i, j in zip(x[\"chr\"], x[\"start\"])]\n",
    "    )\n",
    ")\n",
    "df_covered_repeats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b6db9d7f-bf41-4429-984e-e57f764c8346",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting analysis of sequence: chr1\n",
      "Starting analysis of sequence: chr10\n",
      "Starting analysis of sequence: chr11\n",
      "Starting analysis of sequence: chr12\n",
      "Starting analysis of sequence: chr13\n",
      "Starting analysis of sequence: chr14\n",
      "Starting analysis of sequence: chr15\n",
      "Starting analysis of sequence: chr16\n",
      "Starting analysis of sequence: chr17\n",
      "Starting analysis of sequence: chr18\n",
      "Starting analysis of sequence: chr19\n",
      "Starting analysis of sequence: chr2\n",
      "Starting analysis of sequence: chr20\n",
      "Starting analysis of sequence: chr21\n",
      "Starting analysis of sequence: chr22\n",
      "Starting analysis of sequence: chr3\n",
      "Starting analysis of sequence: chr4\n",
      "Starting analysis of sequence: chr5\n",
      "Starting analysis of sequence: chr6\n",
      "Starting analysis of sequence: chr7\n",
      "Starting analysis of sequence: chr8\n",
      "Starting analysis of sequence: chr9\n",
      "Starting analysis of sequence: chrX\n",
      "Starting analysis of sequence: chrY\n",
      "CPU times: user 1h 9min 13s, sys: 23 s, total: 1h 9min 36s\n",
      "Wall time: 1h 9min 42s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>str_id</th>\n",
       "      <th>region_len_ref</th>\n",
       "      <th>region_len_mat</th>\n",
       "      <th>region_len_pat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_588068</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_589245</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_590659</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_590969</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_591734</td>\n",
       "      <td>18</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695860</th>\n",
       "      <td>chrY_57188868</td>\n",
       "      <td>11</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695861</th>\n",
       "      <td>chrY_57200838</td>\n",
       "      <td>14</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695862</th>\n",
       "      <td>chrY_57201000</td>\n",
       "      <td>20</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695863</th>\n",
       "      <td>chrY_57202381</td>\n",
       "      <td>23</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1695864</th>\n",
       "      <td>chrY_57202848</td>\n",
       "      <td>26</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1695865 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                str_id  region_len_ref region_len_mat region_len_pat\n",
       "0          chr1_588068              12             12             12\n",
       "1          chr1_589245              12             12             12\n",
       "2          chr1_590659              12             12             12\n",
       "3          chr1_590969              16             16             16\n",
       "4          chr1_591734              18             20             20\n",
       "...                ...             ...            ...            ...\n",
       "1695860  chrY_57188868              11           <NA>             11\n",
       "1695861  chrY_57200838              14           <NA>             15\n",
       "1695862  chrY_57201000              20           <NA>             20\n",
       "1695863  chrY_57202381              23           <NA>             23\n",
       "1695864  chrY_57202848              26           <NA>             28\n",
       "\n",
       "[1695865 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "from multicopy_STR_genotyping import str_utils\n",
    "\n",
    "\n",
    "df_strs_haplotypes = {\n",
    "    \"str_id\": [],\n",
    "    \"region_len_ref\": [],\n",
    "    \"region_len_mat\": [],\n",
    "    \"region_len_pat\": [],\n",
    "}\n",
    "\n",
    "for chromosome, data in df_covered_repeats.groupby(\"chr\"):\n",
    "    print(\"Starting analysis of sequence:\", chromosome)\n",
    "    df_chr_mat = df_haplo_mat_indel.query(f\"chr == '{chromosome}'\")\n",
    "    df_chr_pat = df_haplo_pat_indel.query(f\"chr == '{chromosome}'\")\n",
    "    for locus in data.to_dict(orient=\"records\"):\n",
    "        if locus[\"chr\"] == \"chrY\":\n",
    "            mat = pd.NA\n",
    "        else:\n",
    "            mat = str_utils.str_len_from_haplotype(locus, df_chr_mat)\n",
    "            \n",
    "        if locus[\"chr\"] == \"chrX\":\n",
    "            pat = pd.NA\n",
    "        else:\n",
    "            pat = str_utils.str_len_from_haplotype(locus, df_chr_pat)\n",
    "        \n",
    "        df_strs_haplotypes[\"str_id\"].append(locus[\"str_id\"])\n",
    "        df_strs_haplotypes[\"region_len_ref\"].append(locus[\"end\"] - locus[\"start\"] + 1)\n",
    "        df_strs_haplotypes[\"region_len_mat\"].append(mat)\n",
    "        df_strs_haplotypes[\"region_len_pat\"].append(pat)\n",
    "\n",
    "df_strs_haplotypes = pd.DataFrame(df_strs_haplotypes)\n",
    "df_strs_haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f03b5dd1-1834-4a5c-91b3-7fadf48b83b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_strs_haplotypes.to_csv(\"../../data/HG002/variants/HG002_GRCh38_STR_lengthts.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
